// blake3-jit/dist/index.js
//#region src/compress.ts
/**
* BLAKE3 Compression Function - Highly Optimized
*
* Optimization techniques applied (from Fleek Network case study):
* 1. Use 16 SMI variables for state instead of TypedArray
* 2. Use 16 SMI variables for message words
* 3. Fully inlined G function (no function call overhead)
* 4. Use `| 0` for integer coercion (forces V8 to use 32-bit ALU)
* 5. Hardcoded permutation swaps using only 2 temporary variables
* 6. Offset-based access pattern (avoid creating new views)
*
* The compression function takes:
* - cv: 8-word chaining value
* - block: 16-word message block (64 bytes)
* - counter: 64-bit block counter
* - blockLen: number of input bytes in this block
* - flags: domain separation flags
*
* And outputs 8 or 16 words depending on whether this is a root node.
*/
/**
* Compress a single block.
*
* This is the hot path - every optimization matters here.
*
* @param cv - Chaining value array
* @param cvOff - Offset into cv
* @param block - Message block words
* @param blockOff - Offset into block
* @param out - Output array (8 or 16 words)
* @param outOff - Offset into out
* @param full - If true, output all 16 words (for XOF); if false, output 8 words
* @param counter - 64-bit block counter
* @param blockLen - Number of bytes in this block (0-64)
* @param flags - Domain separation flags
*/
function compress(cv, cvOff, block, blockOff, out, outOff, full, counter, blockLen, flags) {
	let m0 = block[blockOff] | 0;
	let m1 = block[blockOff + 1] | 0;
	let m2 = block[blockOff + 2] | 0;
	let m3 = block[blockOff + 3] | 0;
	let m4 = block[blockOff + 4] | 0;
	let m5 = block[blockOff + 5] | 0;
	let m6 = block[blockOff + 6] | 0;
	let m7 = block[blockOff + 7] | 0;
	let m8 = block[blockOff + 8] | 0;
	let m9 = block[blockOff + 9] | 0;
	let m10 = block[blockOff + 10] | 0;
	let m11 = block[blockOff + 11] | 0;
	let m12 = block[blockOff + 12] | 0;
	let m13 = block[blockOff + 13] | 0;
	let m14 = block[blockOff + 14] | 0;
	let m15 = block[blockOff + 15] | 0;
	let s0 = cv[cvOff] | 0;
	let s1 = cv[cvOff + 1] | 0;
	let s2 = cv[cvOff + 2] | 0;
	let s3 = cv[cvOff + 3] | 0;
	let s4 = cv[cvOff + 4] | 0;
	let s5 = cv[cvOff + 5] | 0;
	let s6 = cv[cvOff + 6] | 0;
	let s7 = cv[cvOff + 7] | 0;
	let s8 = 1779033703;
	let s9 = 3144134277;
	let s10 = 1013904242;
	let s11 = 2773480762;
	let s12 = counter | 0;
	let s13 = counter / 4294967296 | 0;
	let s14 = blockLen | 0;
	let s15 = flags | 0;
	s0 = (s0 + s4 | 0) + m0 | 0;
	s12 ^= s0;
	s12 = s12 >>> 16 | s12 << 16;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 12 | s4 << 20;
	s0 = (s0 + s4 | 0) + m1 | 0;
	s12 ^= s0;
	s12 = s12 >>> 8 | s12 << 24;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 7 | s4 << 25;
	s1 = (s1 + s5 | 0) + m2 | 0;
	s13 ^= s1;
	s13 = s13 >>> 16 | s13 << 16;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 12 | s5 << 20;
	s1 = (s1 + s5 | 0) + m3 | 0;
	s13 ^= s1;
	s13 = s13 >>> 8 | s13 << 24;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 7 | s5 << 25;
	s2 = (s2 + s6 | 0) + m4 | 0;
	s14 ^= s2;
	s14 = s14 >>> 16 | s14 << 16;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 12 | s6 << 20;
	s2 = (s2 + s6 | 0) + m5 | 0;
	s14 ^= s2;
	s14 = s14 >>> 8 | s14 << 24;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 7 | s6 << 25;
	s3 = (s3 + s7 | 0) + m6 | 0;
	s15 ^= s3;
	s15 = s15 >>> 16 | s15 << 16;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 12 | s7 << 20;
	s3 = (s3 + s7 | 0) + m7 | 0;
	s15 ^= s3;
	s15 = s15 >>> 8 | s15 << 24;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 7 | s7 << 25;
	s0 = (s0 + s5 | 0) + m8 | 0;
	s15 ^= s0;
	s15 = s15 >>> 16 | s15 << 16;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 12 | s5 << 20;
	s0 = (s0 + s5 | 0) + m9 | 0;
	s15 ^= s0;
	s15 = s15 >>> 8 | s15 << 24;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 7 | s5 << 25;
	s1 = (s1 + s6 | 0) + m10 | 0;
	s12 ^= s1;
	s12 = s12 >>> 16 | s12 << 16;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 12 | s6 << 20;
	s1 = (s1 + s6 | 0) + m11 | 0;
	s12 ^= s1;
	s12 = s12 >>> 8 | s12 << 24;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 7 | s6 << 25;
	s2 = (s2 + s7 | 0) + m12 | 0;
	s13 ^= s2;
	s13 = s13 >>> 16 | s13 << 16;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 12 | s7 << 20;
	s2 = (s2 + s7 | 0) + m13 | 0;
	s13 ^= s2;
	s13 = s13 >>> 8 | s13 << 24;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 7 | s7 << 25;
	s3 = (s3 + s4 | 0) + m14 | 0;
	s14 ^= s3;
	s14 = s14 >>> 16 | s14 << 16;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 12 | s4 << 20;
	s3 = (s3 + s4 | 0) + m15 | 0;
	s14 ^= s3;
	s14 = s14 >>> 8 | s14 << 24;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 7 | s4 << 25;
	{
		const t0 = m0, t1 = m1;
		m0 = m2;
		m2 = m3;
		m3 = m10;
		m10 = m12;
		m12 = m9;
		m9 = m11;
		m11 = m5;
		m5 = t0;
		m1 = m6;
		m6 = m4;
		m4 = m7;
		m7 = m13;
		m13 = m14;
		m14 = m15;
		m15 = m8;
		m8 = t1;
	}
	s0 = (s0 + s4 | 0) + m0 | 0;
	s12 ^= s0;
	s12 = s12 >>> 16 | s12 << 16;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 12 | s4 << 20;
	s0 = (s0 + s4 | 0) + m1 | 0;
	s12 ^= s0;
	s12 = s12 >>> 8 | s12 << 24;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 7 | s4 << 25;
	s1 = (s1 + s5 | 0) + m2 | 0;
	s13 ^= s1;
	s13 = s13 >>> 16 | s13 << 16;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 12 | s5 << 20;
	s1 = (s1 + s5 | 0) + m3 | 0;
	s13 ^= s1;
	s13 = s13 >>> 8 | s13 << 24;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 7 | s5 << 25;
	s2 = (s2 + s6 | 0) + m4 | 0;
	s14 ^= s2;
	s14 = s14 >>> 16 | s14 << 16;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 12 | s6 << 20;
	s2 = (s2 + s6 | 0) + m5 | 0;
	s14 ^= s2;
	s14 = s14 >>> 8 | s14 << 24;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 7 | s6 << 25;
	s3 = (s3 + s7 | 0) + m6 | 0;
	s15 ^= s3;
	s15 = s15 >>> 16 | s15 << 16;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 12 | s7 << 20;
	s3 = (s3 + s7 | 0) + m7 | 0;
	s15 ^= s3;
	s15 = s15 >>> 8 | s15 << 24;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 7 | s7 << 25;
	s0 = (s0 + s5 | 0) + m8 | 0;
	s15 ^= s0;
	s15 = s15 >>> 16 | s15 << 16;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 12 | s5 << 20;
	s0 = (s0 + s5 | 0) + m9 | 0;
	s15 ^= s0;
	s15 = s15 >>> 8 | s15 << 24;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 7 | s5 << 25;
	s1 = (s1 + s6 | 0) + m10 | 0;
	s12 ^= s1;
	s12 = s12 >>> 16 | s12 << 16;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 12 | s6 << 20;
	s1 = (s1 + s6 | 0) + m11 | 0;
	s12 ^= s1;
	s12 = s12 >>> 8 | s12 << 24;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 7 | s6 << 25;
	s2 = (s2 + s7 | 0) + m12 | 0;
	s13 ^= s2;
	s13 = s13 >>> 16 | s13 << 16;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 12 | s7 << 20;
	s2 = (s2 + s7 | 0) + m13 | 0;
	s13 ^= s2;
	s13 = s13 >>> 8 | s13 << 24;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 7 | s7 << 25;
	s3 = (s3 + s4 | 0) + m14 | 0;
	s14 ^= s3;
	s14 = s14 >>> 16 | s14 << 16;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 12 | s4 << 20;
	s3 = (s3 + s4 | 0) + m15 | 0;
	s14 ^= s3;
	s14 = s14 >>> 8 | s14 << 24;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 7 | s4 << 25;
	{
		const t0 = m0, t1 = m1;
		m0 = m2;
		m2 = m3;
		m3 = m10;
		m10 = m12;
		m12 = m9;
		m9 = m11;
		m11 = m5;
		m5 = t0;
		m1 = m6;
		m6 = m4;
		m4 = m7;
		m7 = m13;
		m13 = m14;
		m14 = m15;
		m15 = m8;
		m8 = t1;
	}
	s0 = (s0 + s4 | 0) + m0 | 0;
	s12 ^= s0;
	s12 = s12 >>> 16 | s12 << 16;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 12 | s4 << 20;
	s0 = (s0 + s4 | 0) + m1 | 0;
	s12 ^= s0;
	s12 = s12 >>> 8 | s12 << 24;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 7 | s4 << 25;
	s1 = (s1 + s5 | 0) + m2 | 0;
	s13 ^= s1;
	s13 = s13 >>> 16 | s13 << 16;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 12 | s5 << 20;
	s1 = (s1 + s5 | 0) + m3 | 0;
	s13 ^= s1;
	s13 = s13 >>> 8 | s13 << 24;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 7 | s5 << 25;
	s2 = (s2 + s6 | 0) + m4 | 0;
	s14 ^= s2;
	s14 = s14 >>> 16 | s14 << 16;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 12 | s6 << 20;
	s2 = (s2 + s6 | 0) + m5 | 0;
	s14 ^= s2;
	s14 = s14 >>> 8 | s14 << 24;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 7 | s6 << 25;
	s3 = (s3 + s7 | 0) + m6 | 0;
	s15 ^= s3;
	s15 = s15 >>> 16 | s15 << 16;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 12 | s7 << 20;
	s3 = (s3 + s7 | 0) + m7 | 0;
	s15 ^= s3;
	s15 = s15 >>> 8 | s15 << 24;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 7 | s7 << 25;
	s0 = (s0 + s5 | 0) + m8 | 0;
	s15 ^= s0;
	s15 = s15 >>> 16 | s15 << 16;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 12 | s5 << 20;
	s0 = (s0 + s5 | 0) + m9 | 0;
	s15 ^= s0;
	s15 = s15 >>> 8 | s15 << 24;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 7 | s5 << 25;
	s1 = (s1 + s6 | 0) + m10 | 0;
	s12 ^= s1;
	s12 = s12 >>> 16 | s12 << 16;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 12 | s6 << 20;
	s1 = (s1 + s6 | 0) + m11 | 0;
	s12 ^= s1;
	s12 = s12 >>> 8 | s12 << 24;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 7 | s6 << 25;
	s2 = (s2 + s7 | 0) + m12 | 0;
	s13 ^= s2;
	s13 = s13 >>> 16 | s13 << 16;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 12 | s7 << 20;
	s2 = (s2 + s7 | 0) + m13 | 0;
	s13 ^= s2;
	s13 = s13 >>> 8 | s13 << 24;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 7 | s7 << 25;
	s3 = (s3 + s4 | 0) + m14 | 0;
	s14 ^= s3;
	s14 = s14 >>> 16 | s14 << 16;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 12 | s4 << 20;
	s3 = (s3 + s4 | 0) + m15 | 0;
	s14 ^= s3;
	s14 = s14 >>> 8 | s14 << 24;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 7 | s4 << 25;
	{
		const t0 = m0, t1 = m1;
		m0 = m2;
		m2 = m3;
		m3 = m10;
		m10 = m12;
		m12 = m9;
		m9 = m11;
		m11 = m5;
		m5 = t0;
		m1 = m6;
		m6 = m4;
		m4 = m7;
		m7 = m13;
		m13 = m14;
		m14 = m15;
		m15 = m8;
		m8 = t1;
	}
	s0 = (s0 + s4 | 0) + m0 | 0;
	s12 ^= s0;
	s12 = s12 >>> 16 | s12 << 16;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 12 | s4 << 20;
	s0 = (s0 + s4 | 0) + m1 | 0;
	s12 ^= s0;
	s12 = s12 >>> 8 | s12 << 24;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 7 | s4 << 25;
	s1 = (s1 + s5 | 0) + m2 | 0;
	s13 ^= s1;
	s13 = s13 >>> 16 | s13 << 16;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 12 | s5 << 20;
	s1 = (s1 + s5 | 0) + m3 | 0;
	s13 ^= s1;
	s13 = s13 >>> 8 | s13 << 24;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 7 | s5 << 25;
	s2 = (s2 + s6 | 0) + m4 | 0;
	s14 ^= s2;
	s14 = s14 >>> 16 | s14 << 16;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 12 | s6 << 20;
	s2 = (s2 + s6 | 0) + m5 | 0;
	s14 ^= s2;
	s14 = s14 >>> 8 | s14 << 24;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 7 | s6 << 25;
	s3 = (s3 + s7 | 0) + m6 | 0;
	s15 ^= s3;
	s15 = s15 >>> 16 | s15 << 16;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 12 | s7 << 20;
	s3 = (s3 + s7 | 0) + m7 | 0;
	s15 ^= s3;
	s15 = s15 >>> 8 | s15 << 24;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 7 | s7 << 25;
	s0 = (s0 + s5 | 0) + m8 | 0;
	s15 ^= s0;
	s15 = s15 >>> 16 | s15 << 16;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 12 | s5 << 20;
	s0 = (s0 + s5 | 0) + m9 | 0;
	s15 ^= s0;
	s15 = s15 >>> 8 | s15 << 24;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 7 | s5 << 25;
	s1 = (s1 + s6 | 0) + m10 | 0;
	s12 ^= s1;
	s12 = s12 >>> 16 | s12 << 16;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 12 | s6 << 20;
	s1 = (s1 + s6 | 0) + m11 | 0;
	s12 ^= s1;
	s12 = s12 >>> 8 | s12 << 24;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 7 | s6 << 25;
	s2 = (s2 + s7 | 0) + m12 | 0;
	s13 ^= s2;
	s13 = s13 >>> 16 | s13 << 16;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 12 | s7 << 20;
	s2 = (s2 + s7 | 0) + m13 | 0;
	s13 ^= s2;
	s13 = s13 >>> 8 | s13 << 24;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 7 | s7 << 25;
	s3 = (s3 + s4 | 0) + m14 | 0;
	s14 ^= s3;
	s14 = s14 >>> 16 | s14 << 16;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 12 | s4 << 20;
	s3 = (s3 + s4 | 0) + m15 | 0;
	s14 ^= s3;
	s14 = s14 >>> 8 | s14 << 24;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 7 | s4 << 25;
	{
		const t0 = m0, t1 = m1;
		m0 = m2;
		m2 = m3;
		m3 = m10;
		m10 = m12;
		m12 = m9;
		m9 = m11;
		m11 = m5;
		m5 = t0;
		m1 = m6;
		m6 = m4;
		m4 = m7;
		m7 = m13;
		m13 = m14;
		m14 = m15;
		m15 = m8;
		m8 = t1;
	}
	s0 = (s0 + s4 | 0) + m0 | 0;
	s12 ^= s0;
	s12 = s12 >>> 16 | s12 << 16;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 12 | s4 << 20;
	s0 = (s0 + s4 | 0) + m1 | 0;
	s12 ^= s0;
	s12 = s12 >>> 8 | s12 << 24;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 7 | s4 << 25;
	s1 = (s1 + s5 | 0) + m2 | 0;
	s13 ^= s1;
	s13 = s13 >>> 16 | s13 << 16;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 12 | s5 << 20;
	s1 = (s1 + s5 | 0) + m3 | 0;
	s13 ^= s1;
	s13 = s13 >>> 8 | s13 << 24;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 7 | s5 << 25;
	s2 = (s2 + s6 | 0) + m4 | 0;
	s14 ^= s2;
	s14 = s14 >>> 16 | s14 << 16;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 12 | s6 << 20;
	s2 = (s2 + s6 | 0) + m5 | 0;
	s14 ^= s2;
	s14 = s14 >>> 8 | s14 << 24;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 7 | s6 << 25;
	s3 = (s3 + s7 | 0) + m6 | 0;
	s15 ^= s3;
	s15 = s15 >>> 16 | s15 << 16;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 12 | s7 << 20;
	s3 = (s3 + s7 | 0) + m7 | 0;
	s15 ^= s3;
	s15 = s15 >>> 8 | s15 << 24;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 7 | s7 << 25;
	s0 = (s0 + s5 | 0) + m8 | 0;
	s15 ^= s0;
	s15 = s15 >>> 16 | s15 << 16;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 12 | s5 << 20;
	s0 = (s0 + s5 | 0) + m9 | 0;
	s15 ^= s0;
	s15 = s15 >>> 8 | s15 << 24;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 7 | s5 << 25;
	s1 = (s1 + s6 | 0) + m10 | 0;
	s12 ^= s1;
	s12 = s12 >>> 16 | s12 << 16;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 12 | s6 << 20;
	s1 = (s1 + s6 | 0) + m11 | 0;
	s12 ^= s1;
	s12 = s12 >>> 8 | s12 << 24;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 7 | s6 << 25;
	s2 = (s2 + s7 | 0) + m12 | 0;
	s13 ^= s2;
	s13 = s13 >>> 16 | s13 << 16;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 12 | s7 << 20;
	s2 = (s2 + s7 | 0) + m13 | 0;
	s13 ^= s2;
	s13 = s13 >>> 8 | s13 << 24;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 7 | s7 << 25;
	s3 = (s3 + s4 | 0) + m14 | 0;
	s14 ^= s3;
	s14 = s14 >>> 16 | s14 << 16;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 12 | s4 << 20;
	s3 = (s3 + s4 | 0) + m15 | 0;
	s14 ^= s3;
	s14 = s14 >>> 8 | s14 << 24;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 7 | s4 << 25;
	{
		const t0 = m0, t1 = m1;
		m0 = m2;
		m2 = m3;
		m3 = m10;
		m10 = m12;
		m12 = m9;
		m9 = m11;
		m11 = m5;
		m5 = t0;
		m1 = m6;
		m6 = m4;
		m4 = m7;
		m7 = m13;
		m13 = m14;
		m14 = m15;
		m15 = m8;
		m8 = t1;
	}
	s0 = (s0 + s4 | 0) + m0 | 0;
	s12 ^= s0;
	s12 = s12 >>> 16 | s12 << 16;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 12 | s4 << 20;
	s0 = (s0 + s4 | 0) + m1 | 0;
	s12 ^= s0;
	s12 = s12 >>> 8 | s12 << 24;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 7 | s4 << 25;
	s1 = (s1 + s5 | 0) + m2 | 0;
	s13 ^= s1;
	s13 = s13 >>> 16 | s13 << 16;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 12 | s5 << 20;
	s1 = (s1 + s5 | 0) + m3 | 0;
	s13 ^= s1;
	s13 = s13 >>> 8 | s13 << 24;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 7 | s5 << 25;
	s2 = (s2 + s6 | 0) + m4 | 0;
	s14 ^= s2;
	s14 = s14 >>> 16 | s14 << 16;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 12 | s6 << 20;
	s2 = (s2 + s6 | 0) + m5 | 0;
	s14 ^= s2;
	s14 = s14 >>> 8 | s14 << 24;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 7 | s6 << 25;
	s3 = (s3 + s7 | 0) + m6 | 0;
	s15 ^= s3;
	s15 = s15 >>> 16 | s15 << 16;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 12 | s7 << 20;
	s3 = (s3 + s7 | 0) + m7 | 0;
	s15 ^= s3;
	s15 = s15 >>> 8 | s15 << 24;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 7 | s7 << 25;
	s0 = (s0 + s5 | 0) + m8 | 0;
	s15 ^= s0;
	s15 = s15 >>> 16 | s15 << 16;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 12 | s5 << 20;
	s0 = (s0 + s5 | 0) + m9 | 0;
	s15 ^= s0;
	s15 = s15 >>> 8 | s15 << 24;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 7 | s5 << 25;
	s1 = (s1 + s6 | 0) + m10 | 0;
	s12 ^= s1;
	s12 = s12 >>> 16 | s12 << 16;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 12 | s6 << 20;
	s1 = (s1 + s6 | 0) + m11 | 0;
	s12 ^= s1;
	s12 = s12 >>> 8 | s12 << 24;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 7 | s6 << 25;
	s2 = (s2 + s7 | 0) + m12 | 0;
	s13 ^= s2;
	s13 = s13 >>> 16 | s13 << 16;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 12 | s7 << 20;
	s2 = (s2 + s7 | 0) + m13 | 0;
	s13 ^= s2;
	s13 = s13 >>> 8 | s13 << 24;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 7 | s7 << 25;
	s3 = (s3 + s4 | 0) + m14 | 0;
	s14 ^= s3;
	s14 = s14 >>> 16 | s14 << 16;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 12 | s4 << 20;
	s3 = (s3 + s4 | 0) + m15 | 0;
	s14 ^= s3;
	s14 = s14 >>> 8 | s14 << 24;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 7 | s4 << 25;
	{
		const t0 = m0, t1 = m1;
		m0 = m2;
		m2 = m3;
		m3 = m10;
		m10 = m12;
		m12 = m9;
		m9 = m11;
		m11 = m5;
		m5 = t0;
		m1 = m6;
		m6 = m4;
		m4 = m7;
		m7 = m13;
		m13 = m14;
		m14 = m15;
		m15 = m8;
		m8 = t1;
	}
	s0 = (s0 + s4 | 0) + m0 | 0;
	s12 ^= s0;
	s12 = s12 >>> 16 | s12 << 16;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 12 | s4 << 20;
	s0 = (s0 + s4 | 0) + m1 | 0;
	s12 ^= s0;
	s12 = s12 >>> 8 | s12 << 24;
	s8 = s8 + s12 | 0;
	s4 ^= s8;
	s4 = s4 >>> 7 | s4 << 25;
	s1 = (s1 + s5 | 0) + m2 | 0;
	s13 ^= s1;
	s13 = s13 >>> 16 | s13 << 16;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 12 | s5 << 20;
	s1 = (s1 + s5 | 0) + m3 | 0;
	s13 ^= s1;
	s13 = s13 >>> 8 | s13 << 24;
	s9 = s9 + s13 | 0;
	s5 ^= s9;
	s5 = s5 >>> 7 | s5 << 25;
	s2 = (s2 + s6 | 0) + m4 | 0;
	s14 ^= s2;
	s14 = s14 >>> 16 | s14 << 16;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 12 | s6 << 20;
	s2 = (s2 + s6 | 0) + m5 | 0;
	s14 ^= s2;
	s14 = s14 >>> 8 | s14 << 24;
	s10 = s10 + s14 | 0;
	s6 ^= s10;
	s6 = s6 >>> 7 | s6 << 25;
	s3 = (s3 + s7 | 0) + m6 | 0;
	s15 ^= s3;
	s15 = s15 >>> 16 | s15 << 16;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 12 | s7 << 20;
	s3 = (s3 + s7 | 0) + m7 | 0;
	s15 ^= s3;
	s15 = s15 >>> 8 | s15 << 24;
	s11 = s11 + s15 | 0;
	s7 ^= s11;
	s7 = s7 >>> 7 | s7 << 25;
	s0 = (s0 + s5 | 0) + m8 | 0;
	s15 ^= s0;
	s15 = s15 >>> 16 | s15 << 16;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 12 | s5 << 20;
	s0 = (s0 + s5 | 0) + m9 | 0;
	s15 ^= s0;
	s15 = s15 >>> 8 | s15 << 24;
	s10 = s10 + s15 | 0;
	s5 ^= s10;
	s5 = s5 >>> 7 | s5 << 25;
	s1 = (s1 + s6 | 0) + m10 | 0;
	s12 ^= s1;
	s12 = s12 >>> 16 | s12 << 16;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 12 | s6 << 20;
	s1 = (s1 + s6 | 0) + m11 | 0;
	s12 ^= s1;
	s12 = s12 >>> 8 | s12 << 24;
	s11 = s11 + s12 | 0;
	s6 ^= s11;
	s6 = s6 >>> 7 | s6 << 25;
	s2 = (s2 + s7 | 0) + m12 | 0;
	s13 ^= s2;
	s13 = s13 >>> 16 | s13 << 16;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 12 | s7 << 20;
	s2 = (s2 + s7 | 0) + m13 | 0;
	s13 ^= s2;
	s13 = s13 >>> 8 | s13 << 24;
	s8 = s8 + s13 | 0;
	s7 ^= s8;
	s7 = s7 >>> 7 | s7 << 25;
	s3 = (s3 + s4 | 0) + m14 | 0;
	s14 ^= s3;
	s14 = s14 >>> 16 | s14 << 16;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 12 | s4 << 20;
	s3 = (s3 + s4 | 0) + m15 | 0;
	s14 ^= s3;
	s14 = s14 >>> 8 | s14 << 24;
	s9 = s9 + s14 | 0;
	s4 ^= s9;
	s4 = s4 >>> 7 | s4 << 25;
	if (full) {
		out[outOff + 8] = s8 ^ cv[cvOff];
		out[outOff + 9] = s9 ^ cv[cvOff + 1];
		out[outOff + 10] = s10 ^ cv[cvOff + 2];
		out[outOff + 11] = s11 ^ cv[cvOff + 3];
		out[outOff + 12] = s12 ^ cv[cvOff + 4];
		out[outOff + 13] = s13 ^ cv[cvOff + 5];
		out[outOff + 14] = s14 ^ cv[cvOff + 6];
		out[outOff + 15] = s15 ^ cv[cvOff + 7];
	}
	out[outOff] = s0 ^ s8;
	out[outOff + 1] = s1 ^ s9;
	out[outOff + 2] = s2 ^ s10;
	out[outOff + 3] = s3 ^ s11;
	out[outOff + 4] = s4 ^ s12;
	out[outOff + 5] = s5 ^ s13;
	out[outOff + 6] = s6 ^ s14;
	out[outOff + 7] = s7 ^ s15;
}

//#endregion
//#region src/constants.ts
/**
* BLAKE3 Constants
*
* IV values are the same as SHA-256: first 32 bits of the fractional parts
* of the square roots of the first 8 primes (2..19)
*/
const IV = new Uint32Array([
	1779033703,
	3144134277,
	1013904242,
	2773480762,
	1359893119,
	2600822924,
	528734635,
	1541459225
]);
const CHUNK_START = 1;
const CHUNK_END = 2;
const PARENT = 4;
const ROOT = 8;
const KEYED_HASH = 16;
const DERIVE_KEY_CONTEXT = 32;
const DERIVE_KEY_MATERIAL = 64;
const OUT_LEN = 32;
const KEY_LEN = 32;
const BLOCK_LEN = 64;
const CHUNK_LEN = 1024;
const MAX_DEPTH = 54;
/**
* Flattened permutation table for compress function optimization.
* This enables direct indexed access: PERMUTATIONS[round * 16 + index]
*/
const PERMUTATIONS = new Uint8Array([
	0,
	1,
	2,
	3,
	4,
	5,
	6,
	7,
	8,
	9,
	10,
	11,
	12,
	13,
	14,
	15,
	2,
	6,
	3,
	10,
	7,
	0,
	4,
	13,
	1,
	11,
	12,
	5,
	9,
	14,
	15,
	8,
	3,
	4,
	10,
	12,
	13,
	2,
	7,
	14,
	6,
	5,
	9,
	0,
	11,
	15,
	8,
	1,
	10,
	7,
	12,
	9,
	14,
	3,
	13,
	15,
	4,
	0,
	11,
	2,
	5,
	8,
	1,
	6,
	12,
	13,
	9,
	11,
	15,
	10,
	14,
	8,
	7,
	2,
	5,
	3,
	0,
	1,
	6,
	4,
	9,
	14,
	11,
	5,
	8,
	12,
	15,
	1,
	13,
	3,
	0,
	10,
	2,
	6,
	4,
	7,
	11,
	15,
	5,
	0,
	1,
	9,
	8,
	6,
	14,
	10,
	2,
	12,
	3,
	4,
	7,
	13
]);

//#endregion
//#region src/utils.ts
/**
* BLAKE3 Utility Functions
*
* Optimized for little-endian systems (most user-facing systems).
* BLAKE3 is little-endian friendly - on little-endian systems we can
* create Uint32Array views directly over input buffers.
*/
/**
* Detect system endianness at module load time.
* On little-endian systems, the byte 0x01 will be at index 0.
*/
const IS_LITTLE_ENDIAN = new Uint8Array(new Uint32Array([16909060]).buffer)[0] === 4;
/**
* Read 16 little-endian 32-bit words from a byte array into a Uint32Array.
* This is only needed on big-endian systems.
*
* @param input - Source byte array
* @param offset - Starting byte offset in input
* @param words - Destination Uint32Array (must have at least 16 elements)
*/
function readLittleEndianWordsFull(input, offset, words) {
	for (let i = 0; i < 16; ++i, offset += 4) words[i] = input[offset] | input[offset + 1] << 8 | input[offset + 2] << 16 | input[offset + 3] << 24;
}
/**
* Read a partial block with zero padding.
* Used for the final block when input length is not a multiple of 64.
*
* @param input - Source byte array
* @param offset - Starting byte offset
* @param length - Number of bytes to read (< 64)
* @param words - Destination Uint32Array (must have 16 elements)
*/
function readLittleEndianWordsPartial(input, offset, length, words) {
	words.fill(0);
	const fullWords = length >>> 2;
	let i = 0;
	for (; i < fullWords; ++i, offset += 4) words[i] = input[offset] | input[offset + 1] << 8 | input[offset + 2] << 16 | input[offset + 3] << 24;
	const remaining = length & 3;
	if (remaining > 0) {
		let word = input[offset];
		if (remaining > 1) word |= input[offset + 1] << 8;
		if (remaining > 2) word |= input[offset + 2] << 16;
		words[i] = word;
	}
}
/**
* Write N bytes from 32-bit words to output.
* Used for variable-length output (XOF mode).
*
* @param words - Source Uint32Array
* @param wordOffset - Starting word offset
* @param output - Destination byte array
* @param byteOffset - Starting byte offset in destination
* @param byteCount - Number of bytes to write
*/
function writeLittleEndianBytesPartial(words, wordOffset, output, byteOffset, byteCount) {
	const fullWords = byteCount >>> 2;
	let i = 0;
	for (; i < fullWords; ++i, byteOffset += 4) {
		const w = words[wordOffset + i];
		output[byteOffset] = w & 255;
		output[byteOffset + 1] = w >>> 8 & 255;
		output[byteOffset + 2] = w >>> 16 & 255;
		output[byteOffset + 3] = w >>> 24 & 255;
	}
	const remaining = byteCount & 3;
	if (remaining > 0) {
		const w = words[wordOffset + i];
		output[byteOffset] = w & 255;
		if (remaining > 1) output[byteOffset + 1] = w >>> 8 & 255;
		if (remaining > 2) output[byteOffset + 2] = w >>> 16 & 255;
	}
}
/**
* Encode a UTF-8 string to Uint8Array.
* Used for derive_key context strings.
*/
function encodeUTF8(str) {
	if (typeof TextEncoder !== "undefined") return new TextEncoder().encode(str);
	const bytes = [];
	for (let i = 0; i < str.length; i++) {
		let c = str.charCodeAt(i);
		if (c < 128) bytes.push(c);
		else if (c < 2048) bytes.push(192 | c >> 6, 128 | c & 63);
		else if (c < 55296 || c >= 57344) bytes.push(224 | c >> 12, 128 | c >> 6 & 63, 128 | c & 63);
		else {
			i++;
			c = 65536 + ((c & 1023) << 10 | str.charCodeAt(i) & 1023);
			bytes.push(240 | c >> 18, 128 | c >> 12 & 63, 128 | c >> 6 & 63, 128 | c & 63);
		}
	}
	return new Uint8Array(bytes);
}
/**
* De Bruijn lookup table for O(1) trailing zero count.
* The expression (n & -n) isolates the lowest set bit.
* Multiplying by the De Bruijn constant maps each power of 2 to a unique 5-bit index.
*/
const CTZ32_TABLE = new Uint8Array([
	0,
	1,
	28,
	2,
	29,
	14,
	24,
	3,
	30,
	22,
	20,
	15,
	25,
	17,
	4,
	8,
	31,
	27,
	13,
	23,
	21,
	19,
	16,
	7,
	26,
	12,
	18,
	6,
	11,
	5,
	10,
	9
]);

//#endregion
//#region src/hasher.ts
/**
* BLAKE3 Hasher - Incremental hashing with support for all modes
*
* Supports:
* - Regular hashing
* - Keyed hashing (MAC)
* - Key derivation (derive_key)
* - XOF (eXtendable Output Function) mode
*/
/**
* Output state for XOF (eXtendable Output Function) mode.
* Allows reading arbitrary amounts of output.
*/
var XofReader = class {
	inputCv;
	blockWords;
	counter;
	blockLen;
	flags;
	outputBlock;
	outputBlockOffset;
	constructor(inputCv, blockWords$1, counter, blockLen, flags) {
		this.inputCv = inputCv;
		this.blockWords = blockWords$1;
		this.counter = counter;
		this.blockLen = blockLen;
		this.flags = flags | ROOT;
		this.outputBlock = new Uint32Array(16);
		this.outputBlockOffset = 64;
	}
	/**
	* Read the next `length` bytes of output.
	*/
	read(length) {
		const output = new Uint8Array(length);
		let outputOffset = 0;
		while (outputOffset < length) {
			if (this.outputBlockOffset >= 64) {
				compress(this.inputCv, 0, this.blockWords, 0, this.outputBlock, 0, true, this.counter++, this.blockLen, this.flags);
				this.outputBlockOffset = 0;
			}
			const available = 64 - this.outputBlockOffset;
			const toCopy = Math.min(available, length - outputOffset);
			const wordOffset = this.outputBlockOffset >>> 2;
			if ((this.outputBlockOffset & 3) === 0 && toCopy >= 4) {
				const fullWords = toCopy >>> 2;
				writeLittleEndianBytesPartial(this.outputBlock, wordOffset, output, outputOffset, fullWords << 2);
				const bytesCopied = fullWords << 2;
				outputOffset += bytesCopied;
				this.outputBlockOffset += bytesCopied;
			} else {
				for (let i = 0; i < toCopy; i++) {
					const wordIdx = this.outputBlockOffset + i >>> 2;
					const byteIdx = this.outputBlockOffset + i & 3;
					output[outputOffset + i] = this.outputBlock[wordIdx] >>> (byteIdx << 3) & 255;
				}
				outputOffset += toCopy;
				this.outputBlockOffset += toCopy;
			}
		}
		return output;
	}
};
/**
* Chunk state for processing input data.
* Each chunk is 1024 bytes and produces an 8-word chaining value.
*/
var ChunkState = class {
	chainingValue;
	chunkCounter;
	blockWords;
	blockLen;
	blocksCompressed;
	flags;
	constructor(keyWords, chunkCounter, flags) {
		this.chainingValue = new Uint32Array(keyWords);
		this.chunkCounter = chunkCounter;
		this.blockWords = new Uint32Array(16);
		this.blockLen = 0;
		this.blocksCompressed = 0;
		this.flags = flags;
	}
	/**
	* Get the flags for the current block.
	*/
	startFlag() {
		return this.blocksCompressed === 0 ? CHUNK_START : 0;
	}
	/**
	* Update the chunk state with input data.
	* Returns the number of bytes consumed.
	*/
	update(input, inputOffset, inputLen) {
		let consumed = 0;
		while (inputLen > 0) {
			if (this.blockLen === BLOCK_LEN) {
				compress(this.chainingValue, 0, this.blockWords, 0, this.chainingValue, 0, false, this.chunkCounter, BLOCK_LEN, this.flags | this.startFlag());
				this.blocksCompressed++;
				this.blockLen = 0;
			}
			const want = BLOCK_LEN - this.blockLen;
			const take = Math.min(want, inputLen);
			if (IS_LITTLE_ENDIAN && this.blockLen === 0 && take === BLOCK_LEN) {
				const inputWords = new Uint32Array(input.buffer, input.byteOffset + inputOffset, 16);
				this.blockWords.set(inputWords);
			} else if (this.blockLen === 0 && take === BLOCK_LEN) readLittleEndianWordsFull(input, inputOffset, this.blockWords);
			else for (let i = 0; i < take; i++) {
				const pos = this.blockLen + i;
				const wordIdx = pos >>> 2;
				const byteIdx = pos & 3;
				if (byteIdx === 0) this.blockWords[wordIdx] = input[inputOffset + i];
				else this.blockWords[wordIdx] |= input[inputOffset + i] << (byteIdx << 3);
			}
			this.blockLen += take;
			inputOffset += take;
			inputLen -= take;
			consumed += take;
		}
		return consumed;
	}
	/**
	* Finalize this chunk and return its output.
	* Returns 8 words (chaining value) or 16 words (if root).
	*/
	output() {
		const usedWords = this.blockLen + 3 >>> 2;
		for (let i = usedWords; i < 16; i++) this.blockWords[i] = 0;
		return {
			inputCv: this.chainingValue,
			blockWords: this.blockWords,
			blockLen: this.blockLen,
			counter: this.chunkCounter,
			flags: this.flags | this.startFlag() | CHUNK_END
		};
	}
	/**
	* Get the number of bytes in this chunk.
	*/
	len() {
		return this.blocksCompressed * BLOCK_LEN + this.blockLen;
	}
};
/**
* Main BLAKE3 Hasher class.
*
* Usage:
*   const hasher = new Hasher();
*   hasher.update(data);
*   const hash = hasher.finalize();
*
* Or with chaining:
*   const hash = new Hasher().update(data).finalize();
*/
var Hasher = class Hasher {
	chunkState;
	keyWords;
	cvStack;
	cvStackLen;
	flags;
	/**
	* Create a new Hasher.
	*
	* @param keyWords - Initial key words (IV for regular hashing)
	* @param flags - Domain separation flags
	*/
	constructor(keyWords, flags) {
		this.keyWords = keyWords ? new Uint32Array(keyWords) : new Uint32Array(IV);
		this.flags = flags ?? 0;
		this.chunkState = new ChunkState(this.keyWords, 0, this.flags);
		this.cvStack = new Uint32Array(MAX_DEPTH * 8);
		this.cvStackLen = 0;
	}
	/**
	* Create a new keyed hasher (MAC).
	*
	* @param key - 32-byte key
	*/
	static newKeyed(key) {
		if (key.length !== KEY_LEN) throw new Error(`Key must be ${KEY_LEN} bytes, got ${key.length}`);
		const keyWords = new Uint32Array(8);
		if (IS_LITTLE_ENDIAN) {
			const view = new Uint32Array(key.buffer, key.byteOffset, 8);
			keyWords.set(view);
		} else for (let i = 0; i < 8; i++) {
			const off = i * 4;
			keyWords[i] = key[off] | key[off + 1] << 8 | key[off + 2] << 16 | key[off + 3] << 24;
		}
		return new Hasher(keyWords, KEYED_HASH);
	}
	/**
	* Create a new key derivation hasher.
	*
	* @param context - Context string for domain separation
	*/
	static newDeriveKey(context) {
		const contextBytes = encodeUTF8(context);
		const contextHasher = new Hasher(new Uint32Array(IV), DERIVE_KEY_CONTEXT);
		contextHasher.update(contextBytes);
		const contextKey = new Uint32Array(8);
		const output = contextHasher.finalizeOutput();
		compress(output.inputCv, 0, output.blockWords, 0, contextKey, 0, false, output.counter, output.blockLen, output.flags | ROOT);
		return new Hasher(contextKey, DERIVE_KEY_MATERIAL);
	}
	/**
	* Push a chaining value onto the stack.
	*/
	pushCv(cv, cvOffset) {
		this.cvStack.set(cv.subarray(cvOffset, cvOffset + 8), this.cvStackLen * 8);
		this.cvStackLen++;
	}
	/**
	* Pop a chaining value from the stack.
	*/
	popCv(out, outOffset) {
		this.cvStackLen--;
		out.set(this.cvStack.subarray(this.cvStackLen * 8, (this.cvStackLen + 1) * 8), outOffset);
	}
	/**
	* Add a chunk's chaining value and merge completed subtrees.
	*/
	addChunkCv(newCv, newCvOffset, totalChunks) {
		const parentBlock = new Uint32Array(16);
		const parentCv = new Uint32Array(8);
		while ((totalChunks & 1) === 0) {
			this.popCv(parentBlock, 0);
			parentBlock.set(newCv.subarray(newCvOffset, newCvOffset + 8), 8);
			compress(this.keyWords, 0, parentBlock, 0, parentCv, 0, false, 0, BLOCK_LEN, this.flags | PARENT);
			newCv = parentCv;
			newCvOffset = 0;
			totalChunks >>>= 1;
		}
		this.pushCv(newCv, newCvOffset);
	}
	/**
	* Update the hasher with input data.
	*
	* @param input - Data to hash
	* @returns this (for chaining)
	*/
	update(input) {
		let inputOffset = 0;
		let inputLen = input.length;
		while (inputLen > 0) {
			if (this.chunkState.len() === CHUNK_LEN) {
				const output = this.chunkState.output();
				const chunkCv = new Uint32Array(8);
				compress(output.inputCv, 0, output.blockWords, 0, chunkCv, 0, false, output.counter, output.blockLen, output.flags);
				const totalChunks = this.chunkState.chunkCounter + 1;
				this.addChunkCv(chunkCv, 0, totalChunks);
				this.chunkState = new ChunkState(this.keyWords, totalChunks, this.flags);
			}
			const want = CHUNK_LEN - this.chunkState.len();
			const take = Math.min(want, inputLen);
			this.chunkState.update(input, inputOffset, take);
			inputOffset += take;
			inputLen -= take;
		}
		return this;
	}
	/**
	* Get the output parameters (for XOF mode or finalization).
	*/
	finalizeOutput() {
		let output = this.chunkState.output();
		let parentBlock = new Uint32Array(16);
		let cv = new Uint32Array(8);
		if (this.cvStackLen > 0) {
			compress(output.inputCv, 0, output.blockWords, 0, cv, 0, false, output.counter, output.blockLen, output.flags);
			while (this.cvStackLen > 0) {
				this.cvStackLen--;
				parentBlock.set(this.cvStack.subarray(this.cvStackLen * 8, (this.cvStackLen + 1) * 8), 0);
				parentBlock.set(cv, 8);
				if (this.cvStackLen > 0) compress(this.keyWords, 0, parentBlock, 0, cv, 0, false, 0, BLOCK_LEN, this.flags | PARENT);
				else return {
					inputCv: this.keyWords,
					blockWords: parentBlock,
					blockLen: BLOCK_LEN,
					counter: 0,
					flags: this.flags | PARENT
				};
			}
		}
		return output;
	}
	/**
	* Finalize the hash and return the result.
	*
	* @param outputLength - Number of bytes to output (default: 32)
	* @returns The hash output
	*/
	finalize(outputLength = OUT_LEN) {
		const output = this.finalizeOutput();
		const result = new Uint8Array(outputLength);
		if (outputLength <= 64) {
			const outWords = new Uint32Array(16);
			compress(output.inputCv, 0, output.blockWords, 0, outWords, 0, outputLength > 32, output.counter, output.blockLen, output.flags | ROOT);
			if (IS_LITTLE_ENDIAN) {
				const outBytes = new Uint8Array(outWords.buffer);
				result.set(outBytes.subarray(0, outputLength));
			} else writeLittleEndianBytesPartial(outWords, 0, result, 0, outputLength);
		} else {
			const full = this.finalizeXof().read(outputLength);
			result.set(full);
		}
		return result;
	}
	/**
	* Finalize and return an XOF reader for arbitrary-length output.
	*/
	finalizeXof() {
		const output = this.finalizeOutput();
		return new XofReader(new Uint32Array(output.inputCv), new Uint32Array(output.blockWords), output.counter, output.blockLen, output.flags);
	}
};

//#endregion
//#region src/wasm-simd.ts
/**
* BLAKE3 WASM SIMD - Runtime bytecode generation
*
* Generates WebAssembly SIMD bytecode at runtime to process 4 compress
* operations in parallel using 128-bit SIMD vectors (i32x4).
*
* Key insight: One i32x4.add instruction performs 4 parallel additions,
* giving us 4x throughput for the same number of instructions.
*
* Memory layout (all values are transposed for SIMD access):
*   0-511:    4 x 16 message words (m0_0,m0_1,m0_2,m0_3, m1_0,m1_1,m1_2,m1_3, ...)
*   512-639:  4 x 8 chaining values
*   640-767:  4 x 8 output values
*   768-783:  4 x counter low
*   784-799:  4 x counter high
*   800-815:  4 x block length
*   816-831:  4 x flags
*/
function toLebU32Min2(n) {
	return [n & 127 | 128, n >>> 7 & 127];
}
function toLebU32Padded5(n) {
	return [
		n & 127 | 128,
		n >>> 7 & 127 | 128,
		n >>> 14 & 127 | 128,
		n >>> 21 & 127 | 128,
		n >>> 28 & 15
	];
}
function toSignedLeb128_i32(n) {
	const bytes = [];
	let value = n | 0;
	let more = true;
	while (more) {
		let byte = value & 127;
		value >>= 7;
		if (value === 0 && (byte & 64) === 0 || value === -1 && (byte & 64) !== 0) more = false;
		else byte |= 128;
		bytes.push(byte);
	}
	return bytes;
}
const MSG_ACCESS_ORDER = [
	0,
	1,
	2,
	3,
	4,
	5,
	6,
	7,
	8,
	9,
	10,
	11,
	12,
	13,
	14,
	15,
	2,
	6,
	3,
	10,
	7,
	0,
	4,
	13,
	1,
	11,
	12,
	5,
	9,
	14,
	15,
	8,
	3,
	4,
	10,
	12,
	13,
	2,
	7,
	14,
	6,
	5,
	9,
	0,
	11,
	15,
	8,
	1,
	10,
	7,
	12,
	9,
	14,
	3,
	13,
	15,
	4,
	0,
	11,
	2,
	5,
	8,
	1,
	6,
	12,
	13,
	9,
	11,
	15,
	10,
	14,
	8,
	7,
	2,
	5,
	3,
	0,
	1,
	6,
	4,
	9,
	14,
	11,
	5,
	8,
	12,
	15,
	1,
	13,
	3,
	0,
	10,
	2,
	6,
	4,
	7,
	11,
	15,
	5,
	0,
	1,
	9,
	8,
	6,
	14,
	10,
	2,
	12,
	3,
	4,
	7,
	13
];
/**
* Generate the WASM module bytecode with compress4x, compressChunks4x, and compressParent functions.
*/
function generateWasmBytes() {
	const code = [];
	function put(bytes) {
		code.push(...bytes);
	}
	put([
		0,
		97,
		115,
		109
	]);
	put([
		1,
		0,
		0,
		0
	]);
	put([1]);
	put([4]);
	put([1]);
	put([
		96,
		0,
		0
	]);
	put([2]);
	put([11]);
	put([1]);
	put([
		2,
		106,
		115
	]);
	put([
		3,
		109,
		101,
		109
	]);
	put([
		2,
		0,
		1
	]);
	put([3]);
	put([4]);
	put([3]);
	put([0]);
	put([0]);
	put([0]);
	put([7]);
	put([50]);
	put([3]);
	put([10]);
	put([
		99,
		111,
		109,
		112,
		114,
		101,
		115,
		115,
		52,
		120
	]);
	put([0, 0]);
	put([16]);
	put([
		99,
		111,
		109,
		112,
		114,
		101,
		115,
		115,
		67,
		104,
		117,
		110,
		107,
		115,
		52,
		120
	]);
	put([0, 1]);
	put([14]);
	put([
		99,
		111,
		109,
		112,
		114,
		101,
		115,
		115,
		80,
		97,
		114,
		101,
		110,
		116
	]);
	put([0, 2]);
	put([10]);
	const sectionSizeOffset = code.length;
	put([
		0,
		0,
		0,
		0,
		0
	]);
	put([3]);
	const funcSizeOffset = code.length;
	put([
		0,
		0,
		0,
		0,
		0
	]);
	const funcBodyStart = code.length;
	put([1]);
	put([32, 123]);
	for (let i = 0; i < 16; i++) {
		put([65, ...toLebU32Min2(i * 16)]);
		put([
			253,
			0,
			2,
			0
		]);
		put([33, i]);
	}
	for (let i = 0; i < 8; i++) {
		put([65, ...toLebU32Min2(512 + i * 16)]);
		put([
			253,
			0,
			2,
			0
		]);
		put([33, 16 + i]);
	}
	const IV$1 = [
		1779033703,
		3144134277,
		1013904242,
		2773480762
	];
	for (let i = 0; i < 4; i++) {
		const ivBytes = [];
		for (let j = 0; j < 4; j++) {
			ivBytes.push(IV$1[i] & 255);
			ivBytes.push(IV$1[i] >>> 8 & 255);
			ivBytes.push(IV$1[i] >>> 16 & 255);
			ivBytes.push(IV$1[i] >>> 24 & 255);
		}
		put([
			253,
			12,
			...ivBytes
		]);
		put([33, 24 + i]);
	}
	put([65, ...toLebU32Min2(768)]);
	put([
		253,
		0,
		2,
		0
	]);
	put([33, 28]);
	put([65, ...toLebU32Min2(784)]);
	put([
		253,
		0,
		2,
		0
	]);
	put([33, 29]);
	put([65, ...toLebU32Min2(800)]);
	put([
		253,
		0,
		2,
		0
	]);
	put([33, 30]);
	put([65, ...toLebU32Min2(816)]);
	put([
		253,
		0,
		2,
		0
	]);
	put([33, 31]);
	let msgIdx = 0;
	function g(a, b, c, d) {
		const mx = MSG_ACCESS_ORDER[msgIdx++];
		const my = MSG_ACCESS_ORDER[msgIdx++];
		put([32, 16 + a]);
		put([32, 16 + b]);
		put([
			253,
			174,
			1
		]);
		put([32, mx]);
		put([
			253,
			174,
			1
		]);
		put([33, 16 + a]);
		put([32, 16 + d]);
		put([32, 16 + a]);
		put([253, 81]);
		put([34, 16 + d]);
		put([32, 16 + d]);
		put([
			253,
			13,
			2,
			3,
			0,
			1,
			6,
			7,
			4,
			5,
			10,
			11,
			8,
			9,
			14,
			15,
			12,
			13
		]);
		put([33, 16 + d]);
		put([32, 16 + c]);
		put([32, 16 + d]);
		put([
			253,
			174,
			1
		]);
		put([33, 16 + c]);
		put([32, 16 + b]);
		put([32, 16 + c]);
		put([253, 81]);
		put([34, 16 + b]);
		put([65, 12]);
		put([
			253,
			173,
			1
		]);
		put([32, 16 + b]);
		put([65, 20]);
		put([
			253,
			171,
			1
		]);
		put([253, 80]);
		put([33, 16 + b]);
		put([32, 16 + a]);
		put([32, 16 + b]);
		put([
			253,
			174,
			1
		]);
		put([32, my]);
		put([
			253,
			174,
			1
		]);
		put([33, 16 + a]);
		put([32, 16 + d]);
		put([32, 16 + a]);
		put([253, 81]);
		put([34, 16 + d]);
		put([32, 16 + d]);
		put([
			253,
			13,
			1,
			2,
			3,
			0,
			5,
			6,
			7,
			4,
			9,
			10,
			11,
			8,
			13,
			14,
			15,
			12
		]);
		put([33, 16 + d]);
		put([32, 16 + c]);
		put([32, 16 + d]);
		put([
			253,
			174,
			1
		]);
		put([33, 16 + c]);
		put([32, 16 + b]);
		put([32, 16 + c]);
		put([253, 81]);
		put([34, 16 + b]);
		put([65, 7]);
		put([
			253,
			173,
			1
		]);
		put([32, 16 + b]);
		put([65, 25]);
		put([
			253,
			171,
			1
		]);
		put([253, 80]);
		put([33, 16 + b]);
	}
	for (let round = 0; round < 7; round++) {
		g(0, 4, 8, 12);
		g(1, 5, 9, 13);
		g(2, 6, 10, 14);
		g(3, 7, 11, 15);
		g(0, 5, 10, 15);
		g(1, 6, 11, 12);
		g(2, 7, 8, 13);
		g(3, 4, 9, 14);
	}
	for (let i = 0; i < 8; i++) {
		put([65, ...toLebU32Min2(640 + i * 16)]);
		put([32, 16 + i]);
		put([32, 24 + i]);
		put([253, 81]);
		put([
			253,
			11,
			2,
			0
		]);
	}
	put([11]);
	const funcSizeBytes = toLebU32Padded5(code.length - funcBodyStart);
	for (let i = 0; i < 5; i++) code[funcSizeOffset + i] = funcSizeBytes[i];
	const func1SizeOffset = code.length;
	put([
		0,
		0,
		0,
		0,
		0
	]);
	const func1BodyStart = code.length;
	put(generateCompressChunks4xBody());
	const func1SizeBytes = toLebU32Padded5(code.length - func1BodyStart);
	for (let i = 0; i < 5; i++) code[func1SizeOffset + i] = func1SizeBytes[i];
	const func2SizeOffset = code.length;
	put([
		0,
		0,
		0,
		0,
		0
	]);
	const func2BodyStart = code.length;
	put(generateCompressParentBody());
	const func2SizeBytes = toLebU32Padded5(code.length - func2BodyStart);
	for (let i = 0; i < 5; i++) code[func2SizeOffset + i] = func2SizeBytes[i];
	const sectionSizeBytes = toLebU32Padded5(code.length - sectionSizeOffset - 5);
	for (let i = 0; i < 5; i++) code[sectionSizeOffset + i] = sectionSizeBytes[i];
	return new Uint8Array(code);
}
/**
* Generate compressChunks4x WASM function body.
* Processes all 16 blocks of 4 chunks in a single call.
*/
function generateCompressChunks4xBody() {
	const code = [];
	function put(bytes) {
		code.push(...bytes);
	}
	put([2]);
	put([32, 123]);
	put([1, 127]);
	const BATCH_BLOCK_WORDS = SIMD_MEMORY.BATCH_BLOCK_WORDS;
	const BATCH_CV = SIMD_MEMORY.BATCH_CV;
	const BATCH_COUNTER_LOW = SIMD_MEMORY.BATCH_COUNTER_LOW;
	const BATCH_FLAGS_BASE = SIMD_MEMORY.BATCH_FLAGS_BASE;
	const BATCH_OUTPUT = SIMD_MEMORY.BATCH_OUTPUT;
	const IV$1 = [
		1779033703,
		3144134277,
		1013904242,
		2773480762
	];
	for (let i = 0; i < 8; i++) {
		put([65, ...toLebU32Min2(BATCH_CV + i * 16)]);
		put([
			253,
			0,
			2,
			0
		]);
		put([33, 16 + i]);
	}
	put([65, 0]);
	put([33, 32]);
	put([2, 64]);
	put([3, 64]);
	for (let w = 0; w < 16; w++) {
		put([32, 32]);
		put([65, ...toLebU32Min2(256)]);
		put([108]);
		put([65, ...toLebU32Min2(BATCH_BLOCK_WORDS + w * 16)]);
		put([106]);
		put([
			253,
			0,
			2,
			0
		]);
		put([33, w]);
	}
	for (let i = 0; i < 4; i++) {
		const ivBytes = [];
		for (let j = 0; j < 4; j++) {
			ivBytes.push(IV$1[i] & 255);
			ivBytes.push(IV$1[i] >>> 8 & 255);
			ivBytes.push(IV$1[i] >>> 16 & 255);
			ivBytes.push(IV$1[i] >>> 24 & 255);
		}
		put([
			253,
			12,
			...ivBytes
		]);
		put([33, 24 + i]);
	}
	put([65, ...toLebU32Min2(BATCH_COUNTER_LOW)]);
	put([
		253,
		0,
		2,
		0
	]);
	put([33, 28]);
	put([
		253,
		12,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0,
		0
	]);
	put([33, 29]);
	const blockLen64 = [];
	for (let j = 0; j < 4; j++) blockLen64.push(64, 0, 0, 0);
	put([
		253,
		12,
		...blockLen64
	]);
	put([33, 30]);
	put([65, ...toLebU32Min2(BATCH_FLAGS_BASE)]);
	put([
		253,
		0,
		2,
		0
	]);
	put([32, 32]);
	put([69]);
	put([32, 32]);
	put([65, 15]);
	put([70]);
	put([65, 1]);
	put([116]);
	put([114]);
	put([253, 17]);
	put([253, 80]);
	put([33, 31]);
	let msgIdx = 0;
	function g(a, b, c, d) {
		const mx = MSG_ACCESS_ORDER[msgIdx++];
		const my = MSG_ACCESS_ORDER[msgIdx++];
		put([32, 16 + a]);
		put([32, 16 + b]);
		put([
			253,
			174,
			1
		]);
		put([32, mx]);
		put([
			253,
			174,
			1
		]);
		put([33, 16 + a]);
		put([32, 16 + d]);
		put([32, 16 + a]);
		put([253, 81]);
		put([34, 16 + d]);
		put([32, 16 + d]);
		put([
			253,
			13,
			2,
			3,
			0,
			1,
			6,
			7,
			4,
			5,
			10,
			11,
			8,
			9,
			14,
			15,
			12,
			13
		]);
		put([33, 16 + d]);
		put([32, 16 + c]);
		put([32, 16 + d]);
		put([
			253,
			174,
			1
		]);
		put([33, 16 + c]);
		put([32, 16 + b]);
		put([32, 16 + c]);
		put([253, 81]);
		put([34, 16 + b]);
		put([65, 12]);
		put([
			253,
			173,
			1
		]);
		put([32, 16 + b]);
		put([65, 20]);
		put([
			253,
			171,
			1
		]);
		put([253, 80]);
		put([33, 16 + b]);
		put([32, 16 + a]);
		put([32, 16 + b]);
		put([
			253,
			174,
			1
		]);
		put([32, my]);
		put([
			253,
			174,
			1
		]);
		put([33, 16 + a]);
		put([32, 16 + d]);
		put([32, 16 + a]);
		put([253, 81]);
		put([34, 16 + d]);
		put([32, 16 + d]);
		put([
			253,
			13,
			1,
			2,
			3,
			0,
			5,
			6,
			7,
			4,
			9,
			10,
			11,
			8,
			13,
			14,
			15,
			12
		]);
		put([33, 16 + d]);
		put([32, 16 + c]);
		put([32, 16 + d]);
		put([
			253,
			174,
			1
		]);
		put([33, 16 + c]);
		put([32, 16 + b]);
		put([32, 16 + c]);
		put([253, 81]);
		put([34, 16 + b]);
		put([65, 7]);
		put([
			253,
			173,
			1
		]);
		put([32, 16 + b]);
		put([65, 25]);
		put([
			253,
			171,
			1
		]);
		put([253, 80]);
		put([33, 16 + b]);
	}
	for (let round = 0; round < 7; round++) {
		g(0, 4, 8, 12);
		g(1, 5, 9, 13);
		g(2, 6, 10, 14);
		g(3, 7, 11, 15);
		g(0, 5, 10, 15);
		g(1, 6, 11, 12);
		g(2, 7, 8, 13);
		g(3, 4, 9, 14);
	}
	for (let i = 0; i < 8; i++) {
		put([32, 16 + i]);
		put([32, 24 + i]);
		put([253, 81]);
		put([33, 16 + i]);
	}
	put([32, 32]);
	put([65, 1]);
	put([106]);
	put([34, 32]);
	put([65, 16]);
	put([73]);
	put([13, 0]);
	put([11]);
	put([11]);
	for (let i = 0; i < 8; i++) {
		put([65, ...toLebU32Min2(BATCH_OUTPUT + i * 16)]);
		put([32, 16 + i]);
		put([
			253,
			11,
			2,
			0
		]);
	}
	put([11]);
	return code;
}
/**
* Generate compressParent WASM function body.
* Performs a single parent node compression using scalar i32 operations.
* Reads 16 words from PARENT_BLOCK, writes 8 words to CHUNK_CV.
* Uses IV, counter=0, blockLen=64, flags=PARENT(4).
*/
function generateCompressParentBody() {
	const code = [];
	function put(bytes) {
		code.push(...bytes);
	}
	put([1]);
	put([32, 127]);
	const PARENT_BLOCK_OFFSET = SIMD_MEMORY.PARENT_BLOCK;
	const CHUNK_CV_OFFSET = SIMD_MEMORY.CHUNK_CV;
	const IV$1 = [
		1779033703,
		3144134277,
		1013904242,
		2773480762,
		1359893119,
		2600822924,
		528734635,
		1541459225
	];
	for (let i = 0; i < 16; i++) {
		put([65, ...toLebU32Min2(PARENT_BLOCK_OFFSET + i * 4)]);
		put([
			40,
			2,
			0
		]);
		put([33, i]);
	}
	for (let i = 0; i < 8; i++) {
		put([65, ...toSignedLeb128_i32(IV$1[i])]);
		put([33, 16 + i]);
	}
	for (let i = 0; i < 4; i++) {
		put([65, ...toSignedLeb128_i32(IV$1[i])]);
		put([33, 24 + i]);
	}
	put([65, 0]);
	put([33, 28]);
	put([65, 0]);
	put([33, 29]);
	put([
		65,
		192,
		0
	]);
	put([33, 30]);
	put([65, 4]);
	put([33, 31]);
	function g(a, b, c, d, mx, my) {
		const sa = 16 + a, sb = 16 + b, sc = 16 + c, sd = 16 + d;
		put([32, sa]);
		put([32, sb]);
		put([106]);
		put([32, mx]);
		put([106]);
		put([33, sa]);
		put([32, sd]);
		put([32, sa]);
		put([115]);
		put([65, 16]);
		put([120]);
		put([33, sd]);
		put([32, sc]);
		put([32, sd]);
		put([106]);
		put([33, sc]);
		put([32, sb]);
		put([32, sc]);
		put([115]);
		put([65, 12]);
		put([120]);
		put([33, sb]);
		put([32, sa]);
		put([32, sb]);
		put([106]);
		put([32, my]);
		put([106]);
		put([33, sa]);
		put([32, sd]);
		put([32, sa]);
		put([115]);
		put([65, 8]);
		put([120]);
		put([33, sd]);
		put([32, sc]);
		put([32, sd]);
		put([106]);
		put([33, sc]);
		put([32, sb]);
		put([32, sc]);
		put([115]);
		put([65, 7]);
		put([120]);
		put([33, sb]);
	}
	let msgIdx = 0;
	for (let round = 0; round < 7; round++) {
		g(0, 4, 8, 12, MSG_ACCESS_ORDER[msgIdx], MSG_ACCESS_ORDER[msgIdx + 1]);
		msgIdx += 2;
		g(1, 5, 9, 13, MSG_ACCESS_ORDER[msgIdx], MSG_ACCESS_ORDER[msgIdx + 1]);
		msgIdx += 2;
		g(2, 6, 10, 14, MSG_ACCESS_ORDER[msgIdx], MSG_ACCESS_ORDER[msgIdx + 1]);
		msgIdx += 2;
		g(3, 7, 11, 15, MSG_ACCESS_ORDER[msgIdx], MSG_ACCESS_ORDER[msgIdx + 1]);
		msgIdx += 2;
		g(0, 5, 10, 15, MSG_ACCESS_ORDER[msgIdx], MSG_ACCESS_ORDER[msgIdx + 1]);
		msgIdx += 2;
		g(1, 6, 11, 12, MSG_ACCESS_ORDER[msgIdx], MSG_ACCESS_ORDER[msgIdx + 1]);
		msgIdx += 2;
		g(2, 7, 8, 13, MSG_ACCESS_ORDER[msgIdx], MSG_ACCESS_ORDER[msgIdx + 1]);
		msgIdx += 2;
		g(3, 4, 9, 14, MSG_ACCESS_ORDER[msgIdx], MSG_ACCESS_ORDER[msgIdx + 1]);
		msgIdx += 2;
	}
	for (let i = 0; i < 8; i++) {
		put([65, ...toLebU32Min2(CHUNK_CV_OFFSET + i * 4)]);
		put([32, 16 + i]);
		put([32, 24 + i]);
		put([115]);
		put([
			54,
			2,
			0
		]);
	}
	put([11]);
	return code;
}
let wasmInstance = null;
let wasmMemory = null;
let wasmCompress4x = null;
let wasmCompressChunks4x = null;
let wasmCompressParent = null;
let wasmMemoryView = null;
let wasmMemoryView32 = null;
/**
* Check if WASM SIMD is supported.
*/
function isSimdSupported() {
	try {
		const simdTest = new Uint8Array([
			0,
			97,
			115,
			109,
			1,
			0,
			0,
			0,
			1,
			5,
			1,
			96,
			0,
			1,
			123,
			3,
			2,
			1,
			0,
			10,
			22,
			1,
			20,
			0,
			253,
			12,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			0,
			11
		]);
		return WebAssembly.validate(simdTest);
	} catch {
		return false;
	}
}
/**
* Set up arena views over WASM memory.
* Called after WASM memory is allocated.
*/
function setupArenaViews() {
	if (!wasmMemory) return;
	const buffer = wasmMemory.buffer;
	arenaCvStack = new Uint32Array(buffer, SIMD_MEMORY.CV_STACK, 512);
	arenaParentBlock = new Uint32Array(buffer, SIMD_MEMORY.PARENT_BLOCK, 16);
	arenaChunkCv = new Uint32Array(buffer, SIMD_MEMORY.CHUNK_CV, 8);
	arenaTempCvs = new Uint32Array(buffer, SIMD_MEMORY.TEMP_CVS, 32);
	arenaBatchBlockWords = new Uint32Array(buffer, SIMD_MEMORY.BATCH_BLOCK_WORDS, 256 * 4);
	arenaBatchCv = new Uint32Array(buffer, SIMD_MEMORY.BATCH_CV, 32);
	arenaBatchCounterLow = new Uint32Array(buffer, SIMD_MEMORY.BATCH_COUNTER_LOW, 4);
	arenaBatchFlagsBase = new Uint32Array(buffer, SIMD_MEMORY.BATCH_FLAGS_BASE, 4);
	arenaBatchOutput = new Uint32Array(buffer, SIMD_MEMORY.BATCH_OUTPUT, 32);
}
/**
* Initialize the WASM SIMD module synchronously.
* Call this once before using compress4x.
*/
let cachedWasmBytes = null;
function initSimdSync() {
	if (wasmInstance) return true;
	if (!isSimdSupported()) return false;
	try {
		const wasmBytes = cachedWasmBytes || generateWasmBytes();
		cachedWasmBytes = wasmBytes;
		wasmMemory = new WebAssembly.Memory({ initial: 1 });
		const importObject = { js: { mem: wasmMemory } };
		const module = new WebAssembly.Module(wasmBytes.buffer);
		wasmInstance = new WebAssembly.Instance(module, importObject);
		wasmCompress4x = wasmInstance.exports.compress4x;
		wasmCompressChunks4x = wasmInstance.exports.compressChunks4x;
		wasmCompressParent = wasmInstance.exports.compressParent;
		wasmMemoryView = new Uint8Array(wasmMemory.buffer);
		wasmMemoryView32 = new Uint32Array(wasmMemory.buffer);
		setupArenaViews();
		return true;
	} catch (e) {
		console.warn("Failed to initialize WASM SIMD:", e);
		return false;
	}
}
/**
* Memory offsets for SIMD data layout
*
* WASM Arena Pattern: All working buffers live in WASM memory (64KB page)
* This eliminates JS heap allocations during hashing operations.
*/
const SIMD_MEMORY = {
	BLOCK_WORDS: 0,
	CHAINING_VALUES: 512,
	OUTPUT: 640,
	COUNTER_LOW: 768,
	COUNTER_HIGH: 784,
	BLOCK_LEN: 800,
	FLAGS: 816,
	BATCH_BLOCK_WORDS: 832,
	BATCH_CV: 4928,
	BATCH_COUNTER_LOW: 5056,
	BATCH_FLAGS_BASE: 5072,
	BATCH_OUTPUT: 5088,
	CV_STACK: 5216,
	PARENT_BLOCK: 7264,
	CHUNK_CV: 7328,
	TEMP_CVS: 7360
};
let arenaCvStack = null;
let arenaParentBlock = null;
let arenaChunkCv = null;
let arenaTempCvs = null;
let arenaBatchBlockWords = null;
let arenaBatchCv = null;
let arenaBatchCounterLow = null;
let arenaBatchFlagsBase = null;
let arenaBatchOutput = null;
/**
* Get the WASM memory views for writing input data.
*/
function getSimdMemory() {
	if (!wasmMemoryView || !wasmMemoryView32) return null;
	return {
		view: wasmMemoryView,
		view32: wasmMemoryView32
	};
}
/**
* Get the arena buffers for Merkle tree operations.
* These TypedArray views are backed by WASM memory - zero JS heap allocation.
*/
function getArenaBuffers() {
	if (!arenaCvStack || !arenaParentBlock || !arenaChunkCv || !arenaTempCvs) return null;
	return {
		cvStack: arenaCvStack,
		parentBlock: arenaParentBlock,
		chunkCv: arenaChunkCv,
		tempCvs: arenaTempCvs
	};
}
/**
* Run the compress4x function.
* Data must already be set up in WASM memory.
*/
function runCompress4x() {
	if (!wasmCompress4x) throw new Error("WASM SIMD not initialized. Call initSimdSync() first.");
	wasmCompress4x();
}
/**
* Run the compressChunks4x function.
* Processes 4 full chunks (16 blocks each) in a single WASM call.
* Data must already be set up in batch arena buffers.
*/
function runCompressChunks4x() {
	if (!wasmCompressChunks4x) throw new Error("WASM SIMD not initialized. Call initSimdSync() first.");
	wasmCompressChunks4x();
}
/**
* Run the compressParent function.
* Compresses a parent node: reads 16 words from PARENT_BLOCK, writes 8 words to CHUNK_CV.
* Data must already be set up in arena buffers (PARENT_BLOCK at offset 7264).
* Output is written to CHUNK_CV at offset 7328.
*/
function runCompressParent() {
	if (!wasmCompressParent) throw new Error("WASM SIMD not initialized. Call initSimdSync() first.");
	wasmCompressParent();
}

//#endregion
//#region src/hash.ts
/**
* BLAKE3 Hash Function - Simple one-shot API
*
* This provides a simple hash() function optimized for different input sizes.
* For small inputs, uses pure JS. For large inputs, uses WASM SIMD.
*/
let blockWords = null;
const HYPER_CV_STACK = new Uint32Array(512);
const CV_POOL_SIZE = 64;
const CV_POOL = new Uint32Array(CV_POOL_SIZE * 8);
const CV_VIEWS = [];
for (let i = 0; i < CV_POOL_SIZE; i++) CV_VIEWS.push(CV_POOL.subarray(i * 8, i * 8 + 8));
let simdAvailable = false;
const SIMD_THRESHOLD = 4 * CHUNK_LEN;
/**
* Initialize SIMD synchronously (lazy).
*/
function ensureSimdSync() {
	if (simdAvailable) return true;
	simdAvailable = initSimdSync();
	return simdAvailable;
}
const simdChunkCvs = new Uint32Array(32);
const reusableTempCv = new Uint32Array(8);
const reusableChunkCv = new Uint32Array(8);
const reusablePureParentBlock = new Uint32Array(16);
const reusablePureParentCv = new Uint32Array(8);
const reusableSimdCvs = new Uint32Array(32);
const reusableSimdParentBlock = new Uint32Array(16);
const reusableSimdParentCv = new Uint32Array(8);
const reusableOffsets = new Uint32Array(4);
const reusableCounters = new Uint32Array(4);
const reusableBlockLens = new Uint32Array(4);
const reusableFlags = new Uint32Array(4);
const reusableOut8 = new Uint32Array(8);
const reusableOut8View = new Uint8Array(reusableOut8.buffer, 0, 32);
function copyCV8(src, srcOff, dst, dstOff) {
	dst[dstOff] = src[srcOff];
	dst[dstOff + 1] = src[srcOff + 1];
	dst[dstOff + 2] = src[srcOff + 2];
	dst[dstOff + 3] = src[srcOff + 3];
	dst[dstOff + 4] = src[srcOff + 4];
	dst[dstOff + 5] = src[srcOff + 5];
	dst[dstOff + 6] = src[srcOff + 6];
	dst[dstOff + 7] = src[srcOff + 7];
}
/**
* Transpose 4 blocks (64 bytes each) into SIMD memory layout.
* The SIMD compress4x expects: [m0_0,m0_1,m0_2,m0_3, m1_0,m1_1,m1_2,m1_3, ...]
* where m{i}_{j} is message word i from block j.
*
* OPTIMIZED: Processes all 4 blocks together for each word position,
* writing 4 consecutive u32s at once for better cache locality.
*
* @param inputWords - Pre-created Uint32Array view of input (null if unaligned/non-LE).
*                     Created once per hash call to avoid allocation in hot loop.
*/
function transposeBlocksToSimd(input, offsets, blockLens, mem32, blockCount, inputWords) {
	if (blockCount === 4 && blockLens[0] === 64 && blockLens[1] === 64 && blockLens[2] === 64 && blockLens[3] === 64 && inputWords && offsets[0] % 4 === 0 && offsets[1] % 4 === 0 && offsets[2] % 4 === 0 && offsets[3] % 4 === 0) {
		const wordOff0 = offsets[0] >>> 2;
		const wordOff1 = offsets[1] >>> 2;
		const wordOff2 = offsets[2] >>> 2;
		const wordOff3 = offsets[3] >>> 2;
		for (let w = 0; w < 16; w++) {
			const dstBase = w * 4;
			mem32[dstBase] = inputWords[wordOff0 + w];
			mem32[dstBase + 1] = inputWords[wordOff1 + w];
			mem32[dstBase + 2] = inputWords[wordOff2 + w];
			mem32[dstBase + 3] = inputWords[wordOff3 + w];
		}
		return;
	}
	for (let b = 0; b < blockCount; b++) {
		const len = blockLens[b];
		const off = offsets[b];
		if (len === 64) if (inputWords && off % 4 === 0) {
			const wordOff = off >>> 2;
			for (let w = 0; w < 16; w++) mem32[w * 4 + b] = inputWords[wordOff + w];
		} else for (let w = 0; w < 16; w++) {
			const srcOff = off + w * 4;
			mem32[w * 4 + b] = input[srcOff] | input[srcOff + 1] << 8 | input[srcOff + 2] << 16 | input[srcOff + 3] << 24;
		}
		else if (len === 0) for (let w = 0; w < 16; w++) mem32[w * 4 + b] = 0;
		else for (let w = 0; w < 16; w++) {
			const wordOff = w * 4;
			if (wordOff >= len) mem32[w * 4 + b] = 0;
			else if (wordOff + 4 <= len) {
				const srcOff = off + wordOff;
				mem32[w * 4 + b] = input[srcOff] | input[srcOff + 1] << 8 | input[srcOff + 2] << 16 | input[srcOff + 3] << 24;
			} else {
				let word = 0;
				for (let i = 0; i < len - wordOff; i++) word |= input[off + wordOff + i] << i * 8;
				mem32[w * 4 + b] = word;
			}
		}
	}
	for (let b = blockCount; b < 4; b++) for (let w = 0; w < 16; w++) mem32[w * 4 + b] = 0;
}
/**
* Transpose 4 full chunks (4 × 16 blocks = 64 blocks) into batch SIMD memory.
* This is used for the batched compressChunks4x function that processes
* all 16 blocks in a single WASM call.
*
* Memory layout: BATCH_BLOCK_WORDS has 16 positions, each with 16 v128 values.
* Position p, word w: mem32[(p * 64) + (w * 4) + lane]
*
* OPTIMIZED: Processes all 4 chunks together for each (pos, word) pair,
* writing 4 consecutive u32s at once for better cache locality.
*
* @param input - Input data (must have at least 4 full chunks = 4096 bytes)
* @param chunkOffsets - Starting offsets for each of 4 chunks
* @param mem32 - WASM memory view
* @param inputWords - Pre-created Uint32Array view (null if unaligned)
*/
function transposeBatchToSimd(input, chunkOffsets, mem32, inputWords) {
	const BATCH_BASE = SIMD_MEMORY.BATCH_BLOCK_WORDS / 4;
	const chunk0WordBase = chunkOffsets[0] >>> 2;
	const chunk1WordBase = chunkOffsets[1] >>> 2;
	const chunk2WordBase = chunkOffsets[2] >>> 2;
	const chunk3WordBase = chunkOffsets[3] >>> 2;
	if (inputWords && chunkOffsets[0] % 4 === 0) for (let pos = 0; pos < 16; pos++) {
		const posBase = BATCH_BASE + pos * 64;
		const blockWordOff = pos * 16;
		for (let w = 0; w < 16; w++) {
			const dstBase = posBase + w * 4;
			mem32[dstBase] = inputWords[chunk0WordBase + blockWordOff + w];
			mem32[dstBase + 1] = inputWords[chunk1WordBase + blockWordOff + w];
			mem32[dstBase + 2] = inputWords[chunk2WordBase + blockWordOff + w];
			mem32[dstBase + 3] = inputWords[chunk3WordBase + blockWordOff + w];
		}
	}
	else for (let pos = 0; pos < 16; pos++) {
		const posBase = BATCH_BASE + pos * 64;
		const blockByteOff = pos * 64;
		for (let w = 0; w < 16; w++) {
			const dstBase = posBase + w * 4;
			const wordByteOff = w * 4;
			const off0 = chunkOffsets[0] + blockByteOff + wordByteOff;
			mem32[dstBase] = input[off0] | input[off0 + 1] << 8 | input[off0 + 2] << 16 | input[off0 + 3] << 24;
			const off1 = chunkOffsets[1] + blockByteOff + wordByteOff;
			mem32[dstBase + 1] = input[off1] | input[off1 + 1] << 8 | input[off1 + 2] << 16 | input[off1 + 3] << 24;
			const off2 = chunkOffsets[2] + blockByteOff + wordByteOff;
			mem32[dstBase + 2] = input[off2] | input[off2 + 1] << 8 | input[off2 + 2] << 16 | input[off2 + 3] << 24;
			const off3 = chunkOffsets[3] + blockByteOff + wordByteOff;
			mem32[dstBase + 3] = input[off3] | input[off3 + 1] << 8 | input[off3 + 2] << 16 | input[off3 + 3] << 24;
		}
	}
}
const SIMD_CV_BASE = SIMD_MEMORY.CHAINING_VALUES / 4;
const SIMD_OUT_BASE = SIMD_MEMORY.OUTPUT / 4;
const SIMD_COUNTER_LOW_BASE = SIMD_MEMORY.COUNTER_LOW / 4;
const SIMD_COUNTER_HIGH_BASE = SIMD_MEMORY.COUNTER_HIGH / 4;
const SIMD_BLOCK_LEN_BASE = SIMD_MEMORY.BLOCK_LEN / 4;
const BATCH_CV_BASE = SIMD_MEMORY.BATCH_CV / 4;
const BATCH_COUNTER_LOW_BASE = SIMD_MEMORY.BATCH_COUNTER_LOW / 4;
const BATCH_FLAGS_BASE_OFFSET = SIMD_MEMORY.BATCH_FLAGS_BASE / 4;
const BATCH_OUTPUT_BASE = SIMD_MEMORY.BATCH_OUTPUT / 4;
const batchChunkOffsets = new Uint32Array(4);
const SIMD_FLAGS_BASE = SIMD_MEMORY.FLAGS / 4;
/**
* Set up chaining values in SIMD memory (transposed layout).
* Optimized: unrolled loops for common case of 4 chunks.
* cvs is flat: [cv0_word0..cv0_word7, cv1_word0..cv1_word7, ...]
*/
function setupSimdCvs(cvs, mem32, count) {
	if (count === 4) for (let w = 0; w < 8; w++) {
		const base = SIMD_CV_BASE + w * 4;
		mem32[base] = cvs[w];
		mem32[base + 1] = cvs[8 + w];
		mem32[base + 2] = cvs[16 + w];
		mem32[base + 3] = cvs[24 + w];
	}
	else for (let w = 0; w < 8; w++) {
		const base = SIMD_CV_BASE + w * 4;
		for (let c = 0; c < count; c++) mem32[base + c] = cvs[c * 8 + w];
		for (let c = count; c < 4; c++) mem32[base + c] = 0;
	}
}
/**
* Set up SIMD parameters (counters, flags, block lengths).
*/
function setupSimdParams(mem32, counters, blockLens, flagsArr, count) {
	for (let i = 0; i < count; i++) {
		mem32[SIMD_COUNTER_LOW_BASE + i] = counters[i];
		mem32[SIMD_COUNTER_HIGH_BASE + i] = 0;
		mem32[SIMD_BLOCK_LEN_BASE + i] = blockLens[i];
		mem32[SIMD_FLAGS_BASE + i] = flagsArr[i];
	}
	for (let i = count; i < 4; i++) {
		mem32[SIMD_COUNTER_LOW_BASE + i] = 0;
		mem32[SIMD_COUNTER_HIGH_BASE + i] = 0;
		mem32[SIMD_BLOCK_LEN_BASE + i] = 0;
		mem32[SIMD_FLAGS_BASE + i] = 0;
	}
}
/**
* Read output CVs from SIMD memory (untranspose).
*/
function readSimdOutputCvs(mem32, outputCvs, count) {
	if (count === 4) for (let w = 0; w < 8; w++) {
		const base = SIMD_OUT_BASE + w * 4;
		outputCvs[w] = mem32[base];
		outputCvs[8 + w] = mem32[base + 1];
		outputCvs[16 + w] = mem32[base + 2];
		outputCvs[24 + w] = mem32[base + 3];
	}
	else for (let w = 0; w < 8; w++) {
		const base = SIMD_OUT_BASE + w * 4;
		for (let c = 0; c < count; c++) outputCvs[c * 8 + w] = mem32[base + c];
	}
}
function getBlockWords() {
	if (!blockWords) blockWords = new Uint32Array(16);
	return blockWords;
}
/**
* Hash a single chunk (up to 1024 bytes) with pre-created inputWords view.
* This is the optimized version that avoids creating Uint32Array views per chunk.
* (Fleek optimization Step 8)
*/
function hashChunkWithWords(input, inputWords, inputOffset, inputLen, chunkCounter, flags, cv, cvOffset) {
	reusableTempCv.set(IV);
	const fullBlocks = inputLen >>> 6;
	const remainder = inputLen & 63;
	const chunkWordOffset = inputOffset >>> 2;
	if (inputWords && remainder === 0 && inputLen === CHUNK_LEN) {
		let wordOff = chunkWordOffset;
		compress(reusableTempCv, 0, inputWords, wordOff, reusableTempCv, 0, false, chunkCounter, BLOCK_LEN, flags | CHUNK_START);
		wordOff += 16;
		for (let i = 1; i < 15; i++) {
			compress(reusableTempCv, 0, inputWords, wordOff, reusableTempCv, 0, false, chunkCounter, BLOCK_LEN, flags);
			wordOff += 16;
		}
		compress(reusableTempCv, 0, inputWords, wordOff, reusableTempCv, 0, false, chunkCounter, BLOCK_LEN, flags | CHUNK_END);
		cv.set(reusableTempCv, cvOffset);
		return;
	}
	const totalBlocks = fullBlocks + (remainder > 0 ? 1 : 0);
	const block = getBlockWords();
	for (let blockIdx = 0; blockIdx < totalBlocks; blockIdx++) {
		const isFirst = blockIdx === 0;
		const isLast = blockIdx === totalBlocks - 1;
		const blockStart = blockIdx << 6;
		const blockLen = isLast && remainder > 0 ? remainder : BLOCK_LEN;
		let blockFlags = flags;
		if (isFirst) blockFlags |= CHUNK_START;
		if (isLast) blockFlags |= CHUNK_END;
		if (isLast && remainder > 0) readLittleEndianWordsPartial(input, inputOffset + blockStart, blockLen, block);
		else if (inputWords && chunkWordOffset + (blockStart >>> 2) + 16 <= inputWords.length) {
			compress(reusableTempCv, 0, inputWords, chunkWordOffset + (blockStart >>> 2), reusableTempCv, 0, false, chunkCounter, blockLen, blockFlags);
			continue;
		} else readLittleEndianWordsFull(input, inputOffset + blockStart, block);
		compress(reusableTempCv, 0, block, 0, reusableTempCv, 0, false, chunkCounter, blockLen, blockFlags);
	}
	cv.set(reusableTempCv, cvOffset);
}
/**
* Hash input using pure JavaScript.
* Handles the full Merkle tree construction.
*/
function hashPureJS(input, outputLen) {
	const inputLen = input.length;
	if (inputLen === 0) {
		const block = getBlockWords();
		block.fill(0);
		const out$1 = outputLen === 32 ? reusableOut8 : new Uint32Array(outputLen > 32 ? 16 : 8);
		compress(IV, 0, block, 0, out$1, 0, outputLen > 32, 0, 0, CHUNK_START | CHUNK_END | ROOT);
		if (outputLen === 32 && IS_LITTLE_ENDIAN) return reusableOut8View.slice();
		const result$1 = new Uint8Array(outputLen);
		if (IS_LITTLE_ENDIAN) result$1.set(new Uint8Array(out$1.buffer, 0, outputLen));
		else writeLittleEndianBytesPartial(out$1, 0, result$1, 0, outputLen);
		return result$1;
	}
	const numChunks = Math.ceil(inputLen / CHUNK_LEN);
	if (numChunks === 1) {
		const cv = outputLen === 32 ? reusableOut8 : new Uint32Array(outputLen > 32 ? 16 : 8);
		hashChunkRoot(input, 0, inputLen, 0, 0, cv, outputLen > 32);
		if (outputLen === 32 && IS_LITTLE_ENDIAN) return reusableOut8View.slice();
		const result$1 = new Uint8Array(outputLen);
		if (IS_LITTLE_ENDIAN) result$1.set(new Uint8Array(cv.buffer, 0, outputLen));
		else writeLittleEndianBytesPartial(cv, 0, result$1, 0, outputLen);
		return result$1;
	}
	const stack = HYPER_CV_STACK;
	let stackLen = 0;
	const chunkCv = reusableChunkCv;
	const parentBlock = reusablePureParentBlock;
	const parentCv = reusablePureParentCv;
	let inputWords = null;
	const canUseFastPath = IS_LITTLE_ENDIAN && input.byteOffset % 4 === 0;
	if (canUseFastPath) inputWords = new Uint32Array(input.buffer, input.byteOffset, inputLen >>> 2);
	const fullChunks = inputLen >>> 10;
	const lastChunkLen = inputLen & 1023;
	if (canUseFastPath && inputWords) {
		for (let chunkIdx = 0; chunkIdx < fullChunks; chunkIdx++) {
			chunkCv.set(IV);
			let wordOff = chunkIdx << 8;
			compress(chunkCv, 0, inputWords, wordOff, chunkCv, 0, false, chunkIdx, BLOCK_LEN, CHUNK_START);
			wordOff += 16;
			for (let b = 1; b < 15; b++) {
				compress(chunkCv, 0, inputWords, wordOff, chunkCv, 0, false, chunkIdx, BLOCK_LEN, 0);
				wordOff += 16;
			}
			compress(chunkCv, 0, inputWords, wordOff, chunkCv, 0, false, chunkIdx, BLOCK_LEN, CHUNK_END);
			let totalChunks = chunkIdx + 1;
			let cvSrcOff = 0;
			let cvSrc = chunkCv;
			const isLastChunk = chunkIdx === fullChunks - 1 && lastChunkLen === 0;
			while ((totalChunks & 1) === 0 && stackLen > 0) {
				if (stackLen === 1 && isLastChunk) break;
				stackLen--;
				copyCV8(stack, stackLen * 8, parentBlock, 0);
				copyCV8(cvSrc, cvSrcOff, parentBlock, 8);
				compress(IV, 0, parentBlock, 0, parentCv, 0, false, 0, BLOCK_LEN, PARENT);
				cvSrc = parentCv;
				cvSrcOff = 0;
				totalChunks >>>= 1;
			}
			const stackOff = stackLen * 8;
			copyCV8(cvSrc, cvSrcOff, stack, stackOff);
			stackLen++;
		}
		if (lastChunkLen > 0) {
			hashChunkWithWords(input, inputWords, fullChunks * CHUNK_LEN, lastChunkLen, fullChunks, 0, chunkCv, 0);
			let totalChunks = fullChunks + 1;
			let newCv = chunkCv;
			let newCvOffset = 0;
			while ((totalChunks & 1) === 0 && stackLen > 0) {
				if (stackLen === 1) break;
				stackLen--;
				copyCV8(stack, stackLen * 8, parentBlock, 0);
				copyCV8(newCv, newCvOffset, parentBlock, 8);
				compress(IV, 0, parentBlock, 0, parentCv, 0, false, 0, BLOCK_LEN, PARENT);
				newCv = parentCv;
				newCvOffset = 0;
				totalChunks >>>= 1;
			}
			const pushOff = stackLen * 8;
			copyCV8(newCv, newCvOffset, stack, pushOff);
			stackLen++;
		}
	} else for (let chunkIdx = 0; chunkIdx < numChunks; chunkIdx++) {
		const chunkStart = chunkIdx * CHUNK_LEN;
		const chunkLen = Math.min(CHUNK_LEN, inputLen - chunkStart);
		hashChunkWithWords(input, inputWords, chunkStart, chunkLen, chunkIdx, 0, chunkCv, 0);
		let totalChunks = chunkIdx + 1;
		let newCv = chunkCv;
		let newCvOffset = 0;
		const isLastChunk = chunkIdx === numChunks - 1;
		while ((totalChunks & 1) === 0 && stackLen > 0) {
			if (stackLen === 1 && isLastChunk) break;
			stackLen--;
			copyCV8(stack, stackLen * 8, parentBlock, 0);
			copyCV8(newCv, newCvOffset, parentBlock, 8);
			compress(IV, 0, parentBlock, 0, parentCv, 0, false, 0, BLOCK_LEN, PARENT);
			newCv = parentCv;
			newCvOffset = 0;
			totalChunks >>>= 1;
		}
		const pushOff = stackLen * 8;
		copyCV8(newCv, newCvOffset, stack, pushOff);
		stackLen++;
	}
	while (stackLen > 1) {
		stackLen--;
		const rightOff = stackLen * 8;
		stackLen--;
		copyCV8(stack, stackLen * 8, parentBlock, 0);
		copyCV8(stack, rightOff, parentBlock, 8);
		if (stackLen === 0) {
			const out$1 = outputLen === 32 ? reusableOut8 : new Uint32Array(outputLen > 32 ? 16 : 8);
			compress(IV, 0, parentBlock, 0, out$1, 0, outputLen > 32, 0, BLOCK_LEN, PARENT | ROOT);
			if (outputLen === 32 && IS_LITTLE_ENDIAN) return reusableOut8View.slice();
			const result$1 = new Uint8Array(outputLen);
			if (IS_LITTLE_ENDIAN) result$1.set(new Uint8Array(out$1.buffer, 0, outputLen));
			else writeLittleEndianBytesPartial(out$1, 0, result$1, 0, outputLen);
			return result$1;
		}
		compress(IV, 0, parentBlock, 0, parentCv, 0, false, 0, BLOCK_LEN, PARENT);
		copyCV8(parentCv, 0, stack, stackLen * 8);
		stackLen++;
	}
	const out = outputLen === 32 ? reusableOut8 : new Uint32Array(outputLen > 32 ? 16 : 8);
	const lastBlock = getBlockWords();
	lastBlock.fill(0);
	copyCV8(stack, 0, lastBlock, 0);
	compress(IV, 0, lastBlock, 0, out, 0, outputLen > 32, 0, BLOCK_LEN, ROOT);
	if (outputLen === 32 && IS_LITTLE_ENDIAN) return reusableOut8View.slice();
	const result = new Uint8Array(outputLen);
	if (IS_LITTLE_ENDIAN) result.set(new Uint8Array(out.buffer, 0, outputLen));
	else writeLittleEndianBytesPartial(out, 0, result, 0, outputLen);
	return result;
}
/**
* Hash a single chunk that is also the root (single chunk input).
*/
function hashChunkRoot(input, inputOffset, inputLen, chunkCounter, flags, out, fullOutput) {
	reusableTempCv.set(IV);
	const block = getBlockWords();
	const fullBlocks = inputLen >>> 6;
	const remainder = inputLen & 63;
	const totalBlocks = fullBlocks + (remainder > 0 ? 1 : 0) || 1;
	let inputWords = null;
	if (IS_LITTLE_ENDIAN && (input.byteOffset + inputOffset) % 4 === 0 && inputLen >= 4) inputWords = new Uint32Array(input.buffer, input.byteOffset + inputOffset, inputLen >>> 2);
	for (let blockIdx = 0; blockIdx < totalBlocks; blockIdx++) {
		const isFirst = blockIdx === 0;
		const isLast = blockIdx === totalBlocks - 1;
		const blockStart = blockIdx << 6;
		const blockLen = isLast ? remainder || (inputLen > 0 ? BLOCK_LEN : 0) : BLOCK_LEN;
		let blockFlags = flags;
		if (isFirst) blockFlags |= CHUNK_START;
		if (isLast) blockFlags |= CHUNK_END | ROOT;
		if (isLast && remainder > 0) readLittleEndianWordsPartial(input, inputOffset + blockStart, blockLen, block);
		else if (inputLen === 0) block.fill(0);
		else if (inputWords && (blockStart >>> 2) + 16 <= inputWords.length) {
			compress(reusableTempCv, 0, inputWords, blockStart >>> 2, isLast ? out : reusableTempCv, 0, isLast && fullOutput, chunkCounter, blockLen, blockFlags);
			continue;
		} else readLittleEndianWordsFull(input, inputOffset + blockStart, block);
		compress(reusableTempCv, 0, block, 0, isLast ? out : reusableTempCv, 0, isLast && fullOutput, chunkCounter, blockLen, blockFlags);
	}
}
/**
* Hash using WASM SIMD - processes 4 chunks in parallel.
* Falls back to pure JS if SIMD fails.
*/
function hashSimd(input, outputLen) {
	const mem = getSimdMemory();
	if (!mem) return hashPureJS(input, outputLen);
	const { view32 } = mem;
	const inputLen = input.length;
	const numChunks = Math.ceil(inputLen / CHUNK_LEN);
	if (numChunks < 4) return hashPureJS(input, outputLen);
	const arena = getArenaBuffers();
	const useWasmParent = arena !== null;
	let stack;
	let tempCvs;
	let parentBlock;
	let parentCv;
	if (arena) {
		stack = arena.cvStack;
		tempCvs = arena.tempCvs;
		parentBlock = arena.parentBlock;
		parentCv = arena.chunkCv;
	} else {
		stack = HYPER_CV_STACK;
		tempCvs = reusableSimdCvs;
		parentBlock = reusableSimdParentBlock;
		parentCv = reusableSimdParentCv;
	}
	let stackLen = 0;
	const offsets = reusableOffsets;
	const counters = reusableCounters;
	const blockLens = reusableBlockLens;
	const flagsArr = reusableFlags;
	const inputWords = IS_LITTLE_ENDIAN && input.byteOffset % 4 === 0 ? new Uint32Array(input.buffer, input.byteOffset, input.byteLength >>> 2) : null;
	const numFullChunks = inputLen >>> 10;
	let chunkIdx = 0;
	while (chunkIdx < numChunks) {
		const groupSize = Math.min(4, numChunks - chunkIdx);
		if (groupSize === 4 && chunkIdx + 4 <= numFullChunks) {
			batchChunkOffsets[0] = chunkIdx * CHUNK_LEN;
			batchChunkOffsets[1] = (chunkIdx + 1) * CHUNK_LEN;
			batchChunkOffsets[2] = (chunkIdx + 2) * CHUNK_LEN;
			batchChunkOffsets[3] = (chunkIdx + 3) * CHUNK_LEN;
			transposeBatchToSimd(input, batchChunkOffsets, view32, inputWords);
			for (let w = 0; w < 8; w++) {
				const ivWord = IV[w];
				const base = BATCH_CV_BASE + w * 4;
				view32[base] = ivWord;
				view32[base + 1] = ivWord;
				view32[base + 2] = ivWord;
				view32[base + 3] = ivWord;
			}
			view32[BATCH_COUNTER_LOW_BASE] = chunkIdx;
			view32[BATCH_COUNTER_LOW_BASE + 1] = chunkIdx + 1;
			view32[BATCH_COUNTER_LOW_BASE + 2] = chunkIdx + 2;
			view32[BATCH_COUNTER_LOW_BASE + 3] = chunkIdx + 3;
			view32[BATCH_FLAGS_BASE_OFFSET] = 0;
			view32[BATCH_FLAGS_BASE_OFFSET + 1] = 0;
			view32[BATCH_FLAGS_BASE_OFFSET + 2] = 0;
			view32[BATCH_FLAGS_BASE_OFFSET + 3] = 0;
			runCompressChunks4x();
			for (let w = 0; w < 8; w++) {
				const base = BATCH_OUTPUT_BASE + w * 4;
				tempCvs[w] = view32[base];
				tempCvs[8 + w] = view32[base + 1];
				tempCvs[16 + w] = view32[base + 2];
				tempCvs[24 + w] = view32[base + 3];
			}
		} else {
			for (let g = 0; g < groupSize; g++) {
				const base = g * 8;
				tempCvs[base] = IV[0];
				tempCvs[base + 1] = IV[1];
				tempCvs[base + 2] = IV[2];
				tempCvs[base + 3] = IV[3];
				tempCvs[base + 4] = IV[4];
				tempCvs[base + 5] = IV[5];
				tempCvs[base + 6] = IV[6];
				tempCvs[base + 7] = IV[7];
			}
			for (let blockIdx = 0; blockIdx < 16; blockIdx++) {
				for (let g = 0; g < groupSize; g++) {
					const thisChunkIdx = chunkIdx + g;
					const chunkStart = thisChunkIdx * CHUNK_LEN;
					const chunkLen = Math.min(CHUNK_LEN, inputLen - chunkStart);
					const thisBlockStart = chunkStart + blockIdx * BLOCK_LEN;
					const blockStartInChunk = blockIdx * BLOCK_LEN;
					let thisBlockLen = BLOCK_LEN;
					if (blockStartInChunk >= chunkLen) thisBlockLen = 0;
					else if (blockStartInChunk + BLOCK_LEN > chunkLen) thisBlockLen = chunkLen - blockStartInChunk;
					offsets[g] = thisBlockStart;
					counters[g] = thisChunkIdx;
					let flags = 0;
					if (blockIdx === 0) flags |= CHUNK_START;
					const totalBlocksInChunk = Math.ceil(chunkLen / BLOCK_LEN) || 1;
					if (blockIdx === totalBlocksInChunk - 1) flags |= CHUNK_END;
					blockLens[g] = thisBlockLen;
					flagsArr[g] = flags;
				}
				if (blockLens[0] === 0 && blockLens[1] === 0 && blockLens[2] === 0 && blockLens[3] === 0) continue;
				transposeBlocksToSimd(input, offsets, blockLens, view32, groupSize, inputWords);
				setupSimdCvs(tempCvs, view32, groupSize);
				setupSimdParams(view32, counters, blockLens, flagsArr, groupSize);
				runCompress4x();
				readSimdOutputCvs(view32, simdChunkCvs, groupSize);
				for (let g = 0; g < groupSize; g++) {
					if (blockLens[g] === 0) continue;
					const base = g * 8;
					tempCvs[base] = simdChunkCvs[base];
					tempCvs[base + 1] = simdChunkCvs[base + 1];
					tempCvs[base + 2] = simdChunkCvs[base + 2];
					tempCvs[base + 3] = simdChunkCvs[base + 3];
					tempCvs[base + 4] = simdChunkCvs[base + 4];
					tempCvs[base + 5] = simdChunkCvs[base + 5];
					tempCvs[base + 6] = simdChunkCvs[base + 6];
					tempCvs[base + 7] = simdChunkCvs[base + 7];
				}
			}
		}
		for (let g = 0; g < groupSize; g++) {
			const thisChunkIdx = chunkIdx + g;
			let totalChunks = thisChunkIdx + 1;
			let newCvBase = g * 8;
			let newCvSrc = tempCvs;
			const isLastChunk = thisChunkIdx === numChunks - 1;
			while ((totalChunks & 1) === 0 && stackLen > 0) {
				if (stackLen === 1 && isLastChunk) break;
				stackLen--;
				const stackOff = stackLen * 8;
				copyCV8(stack, stackOff, parentBlock, 0);
				copyCV8(newCvSrc, newCvBase, parentBlock, 8);
				if (useWasmParent) runCompressParent();
				else compress(IV, 0, parentBlock, 0, parentCv, 0, false, 0, BLOCK_LEN, PARENT);
				newCvSrc = parentCv;
				newCvBase = 0;
				totalChunks >>>= 1;
			}
			const pushOff = stackLen * 8;
			copyCV8(newCvSrc, newCvBase, stack, pushOff);
			stackLen++;
		}
		chunkIdx += groupSize;
	}
	while (stackLen > 1) {
		stackLen--;
		const rightOff = stackLen * 8;
		stackLen--;
		const leftOff = stackLen * 8;
		copyCV8(stack, leftOff, parentBlock, 0);
		copyCV8(stack, rightOff, parentBlock, 8);
		if (stackLen === 0) {
			const out = outputLen === 32 ? reusableOut8 : new Uint32Array(outputLen > 32 ? 16 : 8);
			compress(IV, 0, parentBlock, 0, out, 0, outputLen > 32, 0, BLOCK_LEN, PARENT | ROOT);
			if (outputLen === 32 && IS_LITTLE_ENDIAN) return reusableOut8View.slice();
			const result = new Uint8Array(outputLen);
			if (IS_LITTLE_ENDIAN) result.set(new Uint8Array(out.buffer, 0, outputLen));
			else writeLittleEndianBytesPartial(out, 0, result, 0, outputLen);
			return result;
		}
		if (useWasmParent) runCompressParent();
		else compress(IV, 0, parentBlock, 0, parentCv, 0, false, 0, BLOCK_LEN, PARENT);
		copyCV8(parentCv, 0, stack, stackLen * 8);
		stackLen++;
	}
	if (stackLen === 1) {
		const block = getBlockWords();
		block.fill(0);
		copyCV8(stack, 0, block, 0);
		const out = outputLen === 32 ? reusableOut8 : new Uint32Array(outputLen > 32 ? 16 : 8);
		compress(IV, 0, block, 0, out, 0, outputLen > 32, 0, BLOCK_LEN, ROOT);
		if (outputLen === 32 && IS_LITTLE_ENDIAN) return reusableOut8View.slice();
		const result = new Uint8Array(outputLen);
		if (IS_LITTLE_ENDIAN) result.set(new Uint8Array(out.buffer, 0, outputLen));
		else writeLittleEndianBytesPartial(out, 0, result, 0, outputLen);
		return result;
	}
	return hashPureJS(input, outputLen);
}
/**
* Hash input data and return the result.
* Automatically uses WASM SIMD for large inputs when available.
*
* @param input - Data to hash
* @param outputLength - Number of bytes to output (default: 32)
* @returns The hash output
*/
function hash(input, outputLength = OUT_LEN) {
	if (input.length >= SIMD_THRESHOLD && ensureSimdSync()) return hashSimd(input, outputLength);
	return hashPureJS(input, outputLength);
}
/**
* Pre-warm SIMD initialization (call early to avoid latency later).
*/
function warmupSimd() {
	return ensureSimdSync();
}
/**
* Hash input data directly into a caller-provided output buffer.
* Zero-allocation for the common 32-byte case - ideal for performance-critical code.
*
* @param input - Data to hash
* @param output - Pre-allocated output buffer (must be at least outputLength bytes)
* @param outputLength - Number of bytes to output (default: 32, max: output.length)
*/
function hashInto(input, output, outputLength = OUT_LEN) {
	if (output.length < outputLength) throw new Error(`Output buffer too small: ${output.length} < ${outputLength}`);
	if (input.length >= SIMD_THRESHOLD && ensureSimdSync()) {
		hashSimdInto(input, output, outputLength);
		return;
	}
	hashPureJSInto(input, output, outputLength);
}
/**
* Internal: Hash using pure JS, writing directly to output buffer.
*/
function hashPureJSInto(input, output, outputLen) {
	const inputLen = input.length;
	if (inputLen === 0) {
		const block = getBlockWords();
		block.fill(0);
		const out = outputLen <= 32 ? reusableOut8 : new Uint32Array(16);
		compress(IV, 0, block, 0, out, 0, outputLen > 32, 0, 0, CHUNK_START | CHUNK_END | ROOT);
		if (IS_LITTLE_ENDIAN) output.set(new Uint8Array(out.buffer, out.byteOffset, outputLen));
		else writeLittleEndianBytesPartial(out, 0, output, 0, outputLen);
		return;
	}
	if (Math.ceil(inputLen / CHUNK_LEN) === 1) {
		const cv = outputLen <= 32 ? reusableOut8 : new Uint32Array(16);
		hashChunkRoot(input, 0, inputLen, 0, 0, cv, outputLen > 32);
		if (IS_LITTLE_ENDIAN) output.set(new Uint8Array(cv.buffer, cv.byteOffset, outputLen));
		else writeLittleEndianBytesPartial(cv, 0, output, 0, outputLen);
		return;
	}
	const result = hashPureJS(input, outputLen);
	output.set(result);
}
/**
* Internal: Hash using SIMD, writing directly to output buffer.
*/
function hashSimdInto(input, output, outputLen) {
	const result = hashSimd(input, outputLen);
	output.set(result);
}

//#endregion
//#region src/index.ts
/**
* Create a new keyed hasher (MAC).
*
* @param key - 32-byte key
* @returns A new Hasher configured for keyed hashing
*
* @example
* ```typescript
* const key = new Uint8Array(32); // Your 32-byte key
* crypto.getRandomValues(key);
*
* const mac = createKeyed(key)
*   .update(message)
*   .finalize();
* ```
*/
function createKeyed(key) {
	return Hasher.newKeyed(key);
}
/**
* Create a new key derivation hasher.
*
* @param context - Context string for domain separation
* @returns A new Hasher configured for key derivation
*
* @example
* ```typescript
* const derivedKey = createDeriveKey("my-app encryption key v1")
*   .update(inputKeyMaterial)
*   .finalize(32);
* ```
*/
function createDeriveKey(context) {
	return Hasher.newDeriveKey(context);
}
/**
* Create a new regular hasher for incremental hashing.
*
* @returns A new Hasher
*
* @example
* ```typescript
* const hasher = createHasher();
* hasher.update(chunk1);
* hasher.update(chunk2);
* const digest = hasher.finalize();
* ```
*/
function createHasher() {
	return new Hasher();
}
if (typeof globalThis !== "undefined" && typeof globalThis.document !== "undefined") queueMicrotask(() => {
	warmupSimd();
});
var src_default = {
	hash,
	hashInto,
	Hasher,
	createHasher,
	createKeyed,
	createDeriveKey
};

//#endregion
export { Hasher, XofReader, createDeriveKey, createHasher, createKeyed, src_default as default, hash, hashInto, warmupSimd };
